---
title: "NMT-Seq EB: Correlation between expression and methylation"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(data.table))
source("/Users/ricard/NMT-seq_EB/correlations/utils.R")
```

<!-- Define I/O and options -->
```{r echo=FALSE, include=FALSE}

## I/O ##
io   <- list()
io$base_dir   <- "/Users/ricard/data/NMT-seq_EB"
io$in.sample_metadata <- paste0(io$base_dir,"/sample_sheet.csv")
io$annos_dir  <- paste0(io$base_dir, "/features/filt")
io$rna_file   <- paste0(io$base_dir, "/rna/parsed/sceset.rds")
io$met_file   <- paste0(io$base_dir, "/met/parsed/met_data.tsv.gz")
io$outdir <- "/Users/ricard/NMT-seq_EB/correlations/out/metrna"

## Options ##
opts <- list()

# Filtering options
# opts$min.weight     <- 1      # minimum weight per feature
opts$min.weight     <- 3      # minimum weight per feature
opts$min.coverage   <- 0.3    # minimum coverage per feature across samples
opts$min.cdr        <- 0.3    # minimum cellular detection rate per gene (expr)
opts$fraction.genes <- 0.5    # fraction of genes (expr) to keep based on variance
opts$fraction.sites <- 0.5    # fraction of sites (met) to keep based on variance
# opts$min.n          <- 100    # minimum number of features per annotation
opts$min.s          <- 15     # minimum number of samples to do the correlation
opts$gene_window    <- 2e4    # window length for the overlap between genes and features
# opts$gene_window    <- 1e5    # window length for the overlap between genes and features

# Correlation options
opts$cor_samples <- T         # do correlation across samples?
opts$cor_genes <- T           # do correlation across genes?
opts$method <- "pearson"      # correlation type
opts$weight <- TRUE           # weighted correlation? 
opts$threshold_fdr  <- 0.10   # pvalue threshold for significance in FDR
opts$permutation <- T         # do permutation test?
opts$n_perms <- 100           # Number of random permutations

# genomic contexts
opts$annos <- c("prom_2000_2000_cgi","prom_2000_2000_noncgi","super_enhancers","CTCF","genebody","p300","Oct4","Nanog","DHS","activeEnhancers_ENCODE")

opts$anno_names <- c(
  "prom_2000_2000_cgi"="CGI promoters",
  "prom_2000_2000_noncgi"="non-CGI promoters",
  "prom_2000_2000$"="Promoters",
  "activeEnhancers_ENCODE"="Active enhancers",
  "super_enhancers"="Super enhancers",
  "primed_enhancers"="Primed enhancers",
  "exons"="Exons",
  "introns"="Introns",
  "genebody"="Gene body"
)

# cells 
opts$cells <- fread(io$in.sample_metadata, header=T) %>% .[pass_metQC==T & pass_rnaQC==T & method=="NMT",sample]
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$in.sample_metadata, header=T) %>% .[sample%in%opts$cells]
```

<!-- Load RNA expression data -->
```{r load_rna, echo=FALSE, include=FALSE}
# Load SCESet object
sceset <- readRDS(file = io$rna_file)

# Subset cells
sceset <- sceset[,opts$cells]
# sceset <- filter(sceset, culture=="Serum" & type=="NMT")

# Create data.table
rna <- exprs(sceset) %>% t %>% as.data.table(keep.rownames = "sample") %>% melt(id.vars = "sample", value.name = "expr", variable.name = "gene")

# Extract gene coordinates metadata to do the overlap
rna_metadata <- fData(sceset) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% .[,c("chr", "start", "end", "gene", "ens_id")]  %>% 
  .[, chr := as.factor(sub("chr", "", chr))] %>% setnames("ens_id", "id")
```

<!-- Load methylation data -->
```{r load_met, echo=FALSE, include=FALSE}
metrna_dt <- fread(sprintf("zcat < %s", io$met_file)) %>% as.data.table %>% .[anno%in%opts$annos & sample%in%opts$cells] %>% .[,anno:=factor(anno, levels=opts$annos)]
```

<!-- Load genomic contexts -->
```{r}
anno_dt <- lapply(opts$annos, function(anno) fread(sprintf("%s/%s.bed", io$annos_dir, anno), 
  colClasses = c("character", "integer", "integer", "character", "character", "character"))[, c(1, 2, 3, 4, 5, 6)]) %>%  
  rbindlist %>% setnames(c("chr", "start", "end", "strand", "id", "anno")) %>% .[,c("anno", "chr") := list(as.factor(anno), as.factor(chr))]
```

<!-- Associate the non-genic contexts with nearby genes -->
```{r associate, echo=FALSE, warning=FALSE}
metrna_list <- list()
# Iterate over each feature
for (ann in unique(metrna_dt$anno)){
  # Subset corresponding anno
  met_tmp <- metrna_dt[anno == ann, ]
  
  # Non gene-associated feature
  if (all(grepl("ENSMUSG", unique(met_tmp$id)) == FALSE)) {
    # Extract coordiantes for methylation sites and for genes
    feature_metadata_tmp <- anno_dt[anno == ann, c("chr", "start", "end", "id")] %>% setkey(chr, start, end)
    rna_metadata_tmp <- rna_metadata[, c("chr", "start", "end", "gene")] %>%
      .[,c("start", "end") := list(start - opts$gene_window, end + opts$gene_window)] %>% setkey(chr, start, end)
    # Do the overlap
    ov <- foverlaps(feature_metadata_tmp, rna_metadata_tmp, nomatch = 0) %>% .[,c("gene", "id")]
    # Add expression levels
    ov_rna <- merge(rna, ov, by = "gene", allow.cartesian = TRUE)
    # Merge with methylation data
    metrna_list[[ann]] <- merge(ov_rna, met_tmp, by = c("sample", "id")) %>% 
      .[,c("sample", "id", "gene", "anno", "expr", "rate", "weight")]
  }
  # Gene-associated feature
  else if (all(grepl("ENSMUSG", unique(met_tmp$id)) == TRUE)) {
    met_tmp <- merge(met_tmp, rna_metadata[, c("id", "gene")], by = "id")
    metrna_list[[ann]] <- merge(rna, met_tmp, by = c("sample", "gene")) %>%
      .[,c("sample", "id", "gene", "anno", "expr", "rate", "weight")]
  }
}
metrna_dt <- rbindlist(metrna_list)
rm(metrna_list, met_tmp, feature_metadata_tmp, rna_metadata_tmp, ov, ov_rna)
```

<!-- General filters for both correlation types -->
```{r filter_all, echo=FALSE}
# Remove features with low weight
metrna_dt <- metrna_dt[weight >= opts$min.weight]
```

<!-- Filter data to correlate across cells -->
```{r filter_across_samples, echo=FALSE}
if (opts$cor_samples) {
  metrna_filt <- copy(metrna_dt)
  
  # Remove genes/features with constant expression across samples
  keep_var_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[, .(var = var(expr)), by = "gene"] %>% .[var > 2, gene])
  keep_var_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[, .(var = wtd.var(rate, weight)), by = c("id")] %>% .[var > 2, id])
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(., names(.), function(x,y) x[gene %in% keep_var_genes[[y]] & id %in% keep_var_sites[[y]]]) %>% rbindlist
  
  # Filter genes with low cellular detection rate and sites with low coverage across samples
  nsamples <- length(unique(metrna_filt$sample))
  keep_cov_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(cdr = sum(expr > 0) / .N), by = "gene"] %>% .[cdr >= opts$min.cdr, gene])
  keep_cov_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(cov = .N / nsamples), by = c("id")] %>% .[cov >= opts$min.coverage, id])
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(., names(.), function(x,y) x[gene %in% keep_cov_genes[[y]] & id %in% keep_cov_sites[[y]]]) %>% rbindlist
  
  # Filter id-gene pairs with small number of samples
  metrna_filt <- metrna_filt[,n:=.N, by=c("gene","id","anno")] %>% .[n >= opts$min.s] %>% .[,n:=NULL]
  
  # Extract highly variable genes/sites across samples
  keep_hv_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(var = var(expr)), by = "gene"] %>% setorder(-var)  %>% head(n = nrow(.) * opts$fraction.genes) %>% .$gene)
  keep_hv_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(var = wtd.var(rate,weight)), by = c("id")] %>% setorder(-var)  %>% head(n = nrow(.) * opts$fraction.sites) %>% .$id)
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(.,names(.), function(x,y) x[gene %in% keep_hv_genes[[y]] & id %in% keep_hv_sites[[y]]]) %>% rbindlist
  
}
```


<!-- Compute correlations  -->
```{r cor_rate, echo=FALSE, include=FALSE}

# Weighted correlation
if (opts$weight == TRUE) {
  if (opts$method != "pearson") { print("Weighted correlation only supported for pearson"); stop() }
  
  # Correlate rate across samples
  if (opts$cor_samples) {
    cor_samples <- metrna_filt[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("id", "gene", "anno")]
  }
  
  # Correlate rate across genes
  if (opts$cor_genes) {
    cor_genes <- metrna_dt[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("sample", "anno")]
  }

# Unweighted correlation
} else {
  
  # Correlate rate across samples
  if (opts$cor_samples) {
    cor_samples <- metrna_filt[, .(V1 = unlist(cor.test(rate, expr, alternative = "two.sided", method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("id", "gene", "anno")]
  }
  
  # Correlate rate across genes
  if (opts$cor_genes) {
    cor_genes <- metrna_dt[, .(V1 = unlist(cor.test(rate, expr, alternative = "two.sided", method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("sample", "anno")]
  }
}

```

<!-- Parse correlations and compute adjusted p-values across genes (both FDR and Bonferroni) -->
```{r}

# Correlation across cells
if (opts$cor_samples) { 
  cor_samples <- cor_samples %>% .[, para := c("r", "t", "p")] %>% dcast(id + gene + anno ~ para, value.var = "V1") %>%
        .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
        .[, c("log_padj_fdr","log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
        .[, sig := padj_fdr <= opts$threshold_fdr] %>%  setorder(padj_fdr)
}

# Correlation across genes
if (opts$cor_genes) { 
  cor_genes <- cor_genes %>% .[,para := c("r", "t", "p")] %>% dcast(sample + anno ~ para, value.var = "V1") %>% 
        .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
        .[, c("log_padj_fdr", "log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
        .[, sig := padj_fdr <= opts$threshold_fdr] %>% setorder(padj_fdr)
}

```

<!-- Run permutation test for the correlation across genes -->
```{r}
if (opts$cor_genes & opts$permutation) {
  foo <- list()
  # pp_vals <- vector(mode = "numeric", length = length(cor_samples$p))
  # r_vals <- vector(mode = "numeric", length = length(cor_samples$r))
  
  for (k in 1:opts$n_perms){
    print(k)
    
    # Copy original methylation data
    metrna_dt_perm <- copy(metrna_dt)
    
    # Permute gene expression levels
    metrna_dt_perm <- metrna_dt_perm[, expr := sample(expr), by = c("sample","anno")]
    
    # Permute methylation values
    metrna_dt_perm <- metrna_dt_perm[, rate := sample(rate), by = c("sample","anno")]
    
    # Compute correlation across genes
    cor_genes_perm <- metrna_dt_perm[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("sample", "anno")] %>% .[,para := c("r", "t", "p")] %>% dcast(sample + anno ~ para, value.var = "V1")
    
    foo[[k]] <- cor_genes_perm[,perm:=k]
    # For each annotation sort by p-value
    # cor_genes_perm <- cor_genes_perm %>% split(.$anno) %>% map(~ .[,.(anno = anno, p = sort(p))]) %>% rbindlist
    
    # Add the sorted p-values
    # pp_vals <- pp_vals + cor_genes_perm$p
    # r_vals <- r_vals + cor_genes_perm$r
  }
  
  stop()
  # Compute the average p-values
  # pp_vals <- pp_vals / opts$n_perms
  
  # Store them in a data.table for later analysis
  # cor_genes_perm <- cor_genes_perm[, p := pp_vals]
}
```

<!-- Run permutation test for the correlation across cells -->
```{r perm_cor_rate, echo=FALSE, include=FALSE}

if (opts$cor_samples & opts$permutation) {
  pp_vals <- vector(mode = "numeric", length = length(cor_samples$p))
  # For each iteration...
  for (k in 1:opts$n_perms){
    print(k)
    # Copy original methylation data
    metrna_filt_perm <- copy(metrna_filt)
    # Permute gene expression levels
    metrna_filt_perm <- metrna_filt_perm[, expr := sample(expr), by = c("gene", "id", "anno")]
    # Compute correlation across samples
    # cor_samples_perm <- metrna_filt_perm[, .(p = wtd.cor(expr, rate, weight)[, c("p.value")]), by = c("id", "gene", "anno")]
    cor_samples_perm <- metrna_filt_perm[, .(p = cor.test(rate, expr, method=opts$method)[["p.value"]]), by = c("id","gene","anno")]
    # For each annotation sort by p-value
    cor_samples_perm <- cor_samples_perm %>% split(.$anno) %>% map(~ .[,.(anno = anno, p = sort(p))]) %>% rbindlist
    # Add the sorted p-values
    pp_vals <- pp_vals + cor_samples_perm %>% .[, p]
  }
  # Compute the average p-values
  pp_vals <- pp_vals / opts$n_perms
  # Store them in a data.table for later analysis
  cor_samples_perm <- cor_samples_perm[, p := pp_vals]
}

```

<!-- Rename annotations -->
```{r rename, echo=FALSE, include=FALSE}
cor_genes[,anno:=stringr::str_replace_all(cor_genes$anno, opts$anno_names)]
if (opts$cor_samples) {
  cor_samples[,anno:=stringr::str_replace_all(cor_samples$anno, opts$anno_names)]
  if (opts$permutation)
    cor_samples_perm[,anno:=stringr::str_replace_all(cor_samples_perm$anno, opts$anno_names)]
}
```

<!-- Boxplot of correlations between methylation and expression -->
```{r boxplot_expr, echo=FALSE, include=TRUE, warning=FALSE, message=FALSE}

# Boxplot of correlations across samples
if (opts$cor_samples) {
  p <- ggplot(cor_samples, aes(x=anno, y=r)) +
    geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#F8766D") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
    labs(title="Correlation across cells", x="", y="Weighted Pearson correlation coefficient") +
    scale_y_continuous(limits=c(-0.7,0.7)) + coord_flip() + boxplot_theme()
  # print(p)
  # pdf(file=paste0(io$outdir, "/metrna_boxplot_acrosssamples.pdf"), width=8, height=6)
  # print(p)
  # dev.off()
}

# Boxplot of correlations across genes
if (opts$cor_genes) {
  p <- ggplot(cor_genes, aes(x=anno, y=r)) +
    geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#F8766D") +
    geom_hline(yintercept=0, colour="black", linetype="dashed") +
    labs(title="Correlation across genes", x="", y="Weighted Pearson correlation coefficient") +
    coord_flip() + boxplot_theme()
  print(p)
  
  # pdf(file=paste0(io$outdir, "/metrna_boxplot_acrossloci.pdf"), width=8, height=6)
  # print(p)
  # dev.off()
  
  # by lineage
  # tmp <- merge(cor_genes, sample_metadata[,c("sample","lineage")], by="sample")
  # p <- ggplot(tmp, aes(x=anno, y=r)) +
  #   # geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#F8766D") +
  #   geom_boxplot(aes(fill=lineage), alpha=0.8, outlier.shape=NA) +
  #   scale_fill_manual(values=c("orangered1","orangered4")) +
  #   geom_hline(yintercept=0, colour="black", linetype="dashed") +
  #   labs(title="Correlation across genes", x="", y="Weighted Pearson correlation coefficient") +
  #   # scale_y_continuous(limits=c(-0.41, 0.05)) +
  #   coord_flip() + boxplot_theme()
  # print(p)
  # pdf(file=paste0(io$outdir, "/metrna_boxplot_acrossgenes_lineage.pdf"), width=8, height=6)
  # print(p)
  # dev.off()
  
}

```


<!-- Volcano plot of p-values against Pearson's r and QQ-plot of p-values -->
```{r, echo=FALSE}
if (opts$cor_samples & opts$permutation) {
  pp <- qq <- list()
  for (n in unique(cor_samples$anno)) {
    
    # Compute positive and negative hits in terms of correlation
    negative_hits <- cor_samples[anno==n & sig==TRUE & r<0,id]
    positive_hits <- cor_samples[anno==n & sig==TRUE & r>0,id]
    all <- nrow(cor_samples[anno == n,])
    
    # Generate volcano plot  
    pp[[n]] <- gg_volcano_plot(cor_samples[anno == n,], title = n)
    
    # Generate permutation plot
    qq[[n]] <- gg_qqplot(cor_samples[anno == n, ], cor_samples_perm[anno == n]$p, title = "")
    
    # Combine and save
    print(plot_grid(pp[[n]], qq[[n]], labels = c("", ""), label_size = 20, ncol = 2, nrow = 1))
    # pdf(file=paste0(io$outdir,"/metrna_volcano_", n, ".pdf"), width = 14, height = 8, useDingbats = FALSE)
    # grid.arrange(pp[[n]], qq[[n]], ncol=2, top = textGrob(n, gp=gpar(fontsize=29, fontface = "bold")), newpage = TRUE)
    # dev.off()
    
  }
}
```



<!-- Save results -->
```{r}
fwrite(round_df(cor_genes,5), paste0(io$outdir,"/metrna_cor_loci.tsv"), sep="\t")
if (opts$cor_samples) {
  fwrite(round_df(cor_samples,5), paste0(io$outdir,"/metrna_cor_samples.tsv"), sep="\t")
  # if (opts$permutation) {
  #   fwrite(round_df(cor_samples_perm,5), paste0(io$outdir,"/metrna_cor_samples_perm.tsv"), sep="\t")
  # }
}
```




IGNORE BELOW

<!-- <!-- Scatterplot of correlations across cells --> -->
<!-- ```{r scatterplots, echo=FALSE, include=FALSE} -->

<!-- # anno_foo="prom_2000_2000" -->
<!-- # sample_foo="E01" -->
<!-- anno_foo="super_enhancers" -->
<!-- sample_foo="B12" -->
<!-- p <- ggplot(metrna_dt[anno==anno_foo & sample==sample_foo], aes(x=rate,y=expr)) + -->
<!--   geom_point(alpha=0.5) + -->
<!--   stat_smooth(method="lm", color="black", alpha=0.5) + -->
<!--   # facet_wrap(~sample) + -->
<!--   xlab("Methylation rate") + ylab("Expression") + -->
<!--   theme( -->
<!--   plot.title = element_text(size=30, hjust=0.5, margin=margin(0,0,20,0)), -->
<!--   axis.title.y = element_text(colour="black", size=28, vjust=1.5), -->
<!--   axis.title.x = element_text(colour="black", size=28, vjust=1.5, margin=margin(15,0,0,0)), -->
<!--   axis.text.x = element_text(colour="black",size=rel(1.8)), -->
<!--   axis.text.y = element_text(colour="black",size=rel(1.8)), -->
<!--   # axis.line = element_line(colour="black", size=rel(0.7)), -->
<!--   # axis.ticks.x = element_line(colour="black", size=rel(0.8)), -->
<!--   # axis.ticks.y = element_blank(), -->
<!--   legend.position="none", -->
<!--   panel.background = element_blank(), -->
<!--   panel.border = element_blank(), -->
<!--   panel.grid = element_blank() -->
<!--   ) -->

<!-- # pdf(file=paste0("out/scatter_", features[i], ".pdf"), width = 8, height = 8) -->
<!-- # print(p) -->
<!-- # dev.off() -->

<!-- ``` -->