---
title: "NMT-seq ESC: profiles of DNA methylation and accessibility and its relationship with gene expression"
output: 
  BiocStyle::html_document:
    toc: false
    fig_width: 10
    fig_height: 8
---


```{r echo=FALSE, include=FALSE}
library(scater)
library(data.table)
library(purrr)
library(ggplot2)
```

```{r define_fncs, echo=FALSE}
theme_pub <- function() {
    theme(
      plot.margin = unit(c(t=1,r=1,b=1,l=1), "cm"),
      plot.title = element_text(size=25,hjust=0.5),
      axis.text=element_text(size=15, colour="black"),
      axis.title.x=element_text(size=17, margin=margin(10,0,0,0)),
      axis.title.y=element_text(size=17, margin=margin(0,10,0,0)),
      axis.line = element_line(size=rel(1.0)),
      axis.ticks = element_line(size=rel(1.3), color="black"),
      legend.key = element_blank(),
      legend.position = "top",
      legend.direction = "horizontal",
      # legend.key.size= unit(0.5, "cm"),
      legend.key.width=unit(1.2,"line"),
      legend.key.height=unit(1.0,"line"),
      legend.margin = margin(t=10, r=0, b=0, l=0, unit="pt"),
      legend.title = element_blank(),
      legend.text = element_text(size=15),
      panel.border=element_blank(),
      panel.grid.major=element_blank(),
      panel.grid.minor=element_blank(),
      panel.background=element_blank()
    )
}

    # theme(
    #   # plot.margin = unit(c(t=1,r=1,b=1,l=1), "cm"),
    #   plot.title = element_text(size=25,hjust=0.5),
    #   axis.text=element_text(size=rel(1.5), colour="black"),
    #   axis.title.x=element_text(size=20, margin=margin(10,0,0,0)),
    #   axis.title.y=element_text(size=20, margin=margin(0,10,0,0)),
    #   axis.line = element_line(size=rel(1.0)),
    #   axis.ticks = element_line(size=rel(1.2)),
    #   legend.key = element_blank(),
    #   legend.position = "right",
    #   legend.direction = "vertical",
    #   # legend.key.size= unit(0.5, "cm"),
    #   legend.key.width=unit(1.5,"line"),
    #   legend.key.height=unit(1.5,"line"),
    #   # legend.margin = margin(t=10, r=0, b=0, l=0, unit="pt"),
    #   legend.title = element_text(size=16, face="bold"),
    #   legend.text = element_text(size=18),
    #   panel.border=element_blank(),
    #   panel.grid.major=element_blank(),
    #   panel.grid.minor=element_blank(),
    #   panel.background=element_blank()
    # )
```

```{r define_opts, echo=FALSE, include=FALSE}

## I/O ##
io <- list()
if (grepl("ricard",Sys.info()['nodename'])) {
  io$basedir <- "/Users/ricard/data/NMT-seq"
  io$outdir <- "/Users/ricard/NMT-seq/profiles/metaccrna/out"
} else {
  stop()
  # io$basedir <- "/hps/nobackup/stegle/users/ricard/NMT-seq"
  # io$outdir <- "/homes/ricard/NOME-seq/profiles/metaccrna/out"
}
io$features.indir <- paste(io$basedir,"features/filt",sep="/")
io$in.sample_metadata <- paste0(io$basedir,"/sample_info_all.txt")
io$met.indir <- paste(io$basedir,"met/raw/filtered/unstranded/binarised/",sep="/")
io$acc.indir <- paste(io$basedir,"acc/raw/filtered/unstranded/binarised/",sep="/")
io$rna.infile <- paste(io$basedir,"rna/parsed/sceset_all.rds",sep="/")

## Options ##
opts <- list()
opts$window_size <- 1500      # symmetric window size to display (up and downstream)
opts$met.tile <- 50           # window size to calculate methylation rates
opts$acc.tile <- 25           # window size to calculate accessibility rates
opts$gene_window <- 10000     # for the overlap between genes and non-genic genomic contexts

# Define which cells to use
# Loading all cells usually takes too much RAM memory, so we are going to use only a subset of cells with similar mean methylation rate values
# opts$cells <- fread(io$in.sample_metadata, header=T, sep=" ") %>% .[passQC_met==T & passQC_acc==T & passQC_rna==T,sample]
opts$cells <- c("A07","A08","A09","B02","C03","C04","C09","D07","D08","E03","F05","F08","G01","G03","G06","H02","H03","H05")
# opts$cells <- c("A07","A08","A09")

# Define which annotations to use and where to center the window (start, end or center)
opts$annotations <- c(
  # "prom_1000_1000"="center",
  # "prom_1000_1000_cgi"="center",
  # "prom_1000_1000_noncgi"="center",
  # "active_enhancers"="center",
  "super_enhancers"="center", 
  "CTCF"="center",
  "p300"="center",
  "Nanog"="center"
  # "Oct4"="center"
  )
```

```{r print_opts, echo=FALSE, include=TRUE}
# cat(sprintf("- Cells: %s\n",paste(opts$cells,collapse=" ")))
# cat(sprintf("- Window size: %d\n",opts$window_size))
# cat(sprintf("- Tile: %d\n",opts$tile))
# cat(sprintf("- Gene extension upstream and downstream for the overlap: %d\n",opts$gene_window))
```

<!-- Load genomic contexts and define windows -->
```{r load_features, echo=FALSE, include=FALSE}
anno_list <- list()
for (anno in names(opts$annotations)) {
  tmp <- fread(sprintf("%s/%s.bed",io$features.indir,anno))[,c(1,2,3,4,5,6)]
  colnames(tmp) <- c("chr","start","end","strand","id","anno")
  
  # Define central position for the window approach
  if (opts$annotations[anno] == "start") {
    tmp <- rbind(tmp[strand=="+",.(chr,start,strand,id,anno)] %>% .[,center:=start] %>% .[,c("start"):=NULL], 
                 tmp[strand=="-",.(chr,end,strand,id,anno)] %>% .[,center:=end] %>% .[,c("end"):=NULL]) 
  }
  if (opts$annotations[anno] == "center") {
    stopifnot(all(tmp[,end] > tmp[,start]))
    tmp <- tmp[,.(chr,start,end,strand,id,anno)][,center:=round(end+start)/2][,c("start","end"):=NULL]
  }
  if (opts$annotations[anno] == "end") {
    tmp <- rbind(tmp[strand=="+",.(chr,end,strand,id,anno)][,center:=end][,c("end"):=NULL], 
                 tmp[strand=="-",.(chr,start,strand,id,anno)][,center:=start][,c("start"):=NULL])
  }
  anno_list[[anno]] <- tmp %>% .[, c("start","end") := list(center-opts$window_size,center+opts$window_size)]
}

anno_df <- rbindlist(anno_list) %>% 
  .[,c("anno","chr"):=list(as.factor(anno),as.factor(sub("chr","",chr)))] %>%
  setkey(chr,start,end)
rm(anno_list)
```

<!-- Load RNA expression data -->
```{r load_expr, echo=FALSE, include=FALSE}

# Load SCEset object
rna <- readRDS(io$rna.infile) 

# Remove duplicated ensembl IDs and spike ins
rna <- rna[!duplicated(fData(rna)$ens_id),]
rna <- rna[!is.na(fData(rna)$ens_id),]

# Extract feature metadata to do the overlap
rna_metadata <- fData(rna) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% 
  .[,c("chr","start","end","gene","ens_id")]  %>% 
  .[,chr:=as.factor(sub("chr","",chr))] %>% 
  setnames("ens_id","id")

# Create data.table from SCEset object
rna <- exprs(rna) %>% t %>% as.data.table(keep.rownames="sample") %>% melt(id.vars="sample", value.name="expr", variable.name="gene")
```

<!-- Load methylation and accessibility data -->
```{r load_accmet, echo=FALSE, include=FALSE}

# Read accessibility data
acc_list <- list()
for (cell in opts$cells) {
  tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$acc.indir,cell), sep="\t", showProgress=F) %>%
    setnames(c("chr","pos","rate")) %>% .[,sample:=cell] %>% .[,c("start","end"):=list(pos,pos)] %>% setnames("pos","bp") %>% 
    setkey("chr","start","end") %>%
    foverlaps(.,anno_df, nomatch=0) %>% .[, c("chr","i.start","i.end") := NULL] %>%
    .[,dist:=ifelse(strand %in% c("+","*"),bp-center,center-bp)] %>% 
    .[, dist:=opts$acc.tile*round(dist/opts$acc.tile)] %>%
    .[,list(rate=mean(rate), n=.N),by=.(sample,id,dist,anno)]
  acc_list[[cell]] <- tmp
}
acc <- rbindlist(acc_list) %>% .[,context:="GC"]
rm(acc_list)

# Read methylation data
met_list <- list()
for (cell in opts$cells) {
  tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$met.indir,cell), sep="\t", showProgress=F) %>%
    setnames(c("chr","pos","rate")) %>% .[,sample:=cell] %>% .[,c("start","end"):=list(pos,pos)] %>% setnames("pos","bp") %>% setkey("chr","start","end") %>%
    foverlaps(.,anno_df, nomatch=0) %>% .[, c("chr","i.start","i.end") := NULL] %>%
    .[,dist:=ifelse(strand %in% c("+","*"),bp-center,center-bp)] %>% 
    .[, dist:=opts$met.tile*round(dist/opts$met.tile)] %>%
    .[,list(rate=mean(rate), n=.N),by=.(sample,id,dist,anno)]
  met_list[[cell]] <- tmp
}
met <- rbindlist(met_list) %>% .[,context:="CG"]
rm(met_list)

# Combine met and acc and remove old data to free some space
metacc <- rbind(met,acc)
rm(acc,met)

```

<!-- Associate the non-genic contexts with the expression of nearby genes -->
```{r overlap, echo=FALSE, include=FALSE}

metaccrna_list <- list()
for (ann in names(opts$annotations)) {
  
  # Subset corresponding anno
  metacc_tmp <- metacc[anno==ann,]
  
  # Non gene-associated feature (enhancers, TF binding sites, etc.)
  if (all(grepl("ENSM",unique(metacc_tmp$id))==FALSE)) {
    # extract coordinates for methylation sites and for genes
    features_tmp <- anno_df[anno==ann,c("chr","start","end","id")] %>% setkey(chr,start,end)
    genes_tmp <- rna_metadata %>% copy %>% .[,c("start","end","id"):=list(start-opts$gene_window,end+opts$gene_window,NULL)] %>% setkey(chr,start,end)
    # do the overlap
    ov <- foverlaps(features_tmp, genes_tmp, nomatch=0) %>% .[,c("gene","id")]
    # add expression levels
    ov_rna <- merge(rna,ov,by="gene",allow.cartesian=TRUE)
    # merge with methylation/accessibility data
    metaccrna_list[[ann]] <- merge(ov_rna,metacc_tmp,by=c("sample","id")) %>%
      .[,c("sample","id","gene","anno","expr","rate","n","dist","context")]
  }
  
  # Gene-associated feature (promoter, genebody, etc.)
  else if (all(grepl("ENSM",unique(metacc_tmp$id))==T)) {
    metacc_tmp <- merge(metacc_tmp, rna_metadata, by="id")
    metaccrna_list[[ann]] <- merge(rna,metacc_tmp,by=c("sample","gene")) %>%
      .[,c("sample","id","gene","anno","expr","rate","n","dist","context")]
  }
}  
metaccrna <- rbindlist(metaccrna_list) %>% droplevels()
```

<!-- Perform basic filtering on the data -->
<!-- ```{r filtering, echo=FALSE, include=FALSE} -->
<!-- opts$min.weight <- 2 -->
<!-- metaccrna_filt = metaccrna[n>=opts$min.weight] -->
<!-- rm(metaccrna) -->
<!-- ``` -->

<!-- Split genes in each group by high, medium and low expression -->
```{r split_expr, echo=FALSE, include=FALSE}
low_expr <- 1
high_expr <- 6

to.plot <- metaccrna %>%
  .[expr<=low_expr, expr_group:="Low"] %>% 
  .[expr>=high_expr, expr_group:="High"] %>%
  .[expr<high_expr & expr>low_expr, expr_group:="Medium"] %>% 
  .[!is.na(expr_group),] %>% 
  .[,expr_group:=factor(expr_group, levels=c("Low","Medium","High"))]

# metaccrna_filt[,list(expr_group=unique(expr_group)),by=c("id","anno")] %>% 
  # split(.$anno) %>% map(~ .[,.N,by="expr_group"]) %>% map(print)
```

<!-- Rename annotations for plotting -->
```{r preproc_data, echo=FALSE, include=FALSE, warning=FALSE}

anno_names <- c(
  "prom_1000_1000_cgi"="CGI promoters",
  "prom_1000_1000_noncgi"="non-CGI promoters",
  "prom_1000_1000_active"="Active promoters",
  "prom_1000_1000_inactive"="Inactive promoters",
  "prom_1000_1000_poised"="Poised promoters",
  "prom_1000_1000$"="Promoters",
  "active_enhancers"="Active enhancers",
  "super_enhancers"="Super enhancers",
  "primed_enhancers"="Primed enhancers"
  )

to.plot$anno <- stringr::str_replace_all(to.plot$anno,anno_names)
```

<!-- Generate stratified profiles -->
```{r plot, echo=FALSE, include=TRUE}

f <- function(x) { return(data.frame(y=mean(x), ymin=mean(x)-sd(x), ymax=mean(x)+sd(x))) }

# to.plot[,tmp:=paste(context,expr_group,sep="_")] %>% .[,tmp:=factor(tmp, levels=c("CG_Low","CG_Medium","CG_High","GC_Low","GC_Medium","GC_High"))]

p_list <- list()
for (ann in unique(to.plot$anno)) {
  p <- ggplot(to.plot[anno==ann,], aes(x=dist, y=rate)) + 
    ggtitle(ann) +
    stat_summary(aes(group=interaction(expr_group,context), colour=context, fill=context, linetype=expr_group), fun.data=mean_se, geom="smooth", alpha=0.2, size=1.3) +
    xlab("Genomic distance from TSS") +
    ylab("Methylation/Accessibility rate") +
    scale_linetype_manual(labels=c("Low","Medium","High"), values=c("dotted","dashed","solid")) +
    scale_color_manual(labels=c("CG methylation","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    scale_fill_manual(labels=c("CG methylation","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    # scale_color_manual(labels=c(" expression (CG)","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    guides(
      colour=guide_legend(title="Context" ),
      linetype=guide_legend(override.aes=list(color="black"), title="Expression group"),
      fill=FALSE
    ) + theme_pub()

  p_list[[ann]] <- p
  
  # pdf(sprintf("%s/%s.pdf",io$outdir,ann), width=10, height=7, useDingbats=F)
  print(p)
  # dev.off()
}
```


```{r, echo=FALSE, include=FALSE}
pp <- cowplot::plot_grid(p_list[["p300"]],p_list[["Nanog"]],p_list[["CTCF"]],p_list[["Super enhancers"]], ncol=2, nrow=2, scale=0.97)
pdf(file=paste0(io$outdir,"/foo.pdf"), width=12, height=12, useDingbats=F)
print(pp)
dev.off()
```

