---
title: "NMT-Seq: Correlation between expression and accessibility"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(plyr))
suppressPackageStartupMessages(library(truncnorm))
```

```{r echo=FALSE, include=FALSE}
# Define ggplot2 theme for scatter plots
scatter_theme <- function(){
  p <- theme(
      plot.title=element_text(size=28, face='bold', margin=margin(0,0,10,0), hjust=0.5),
      plot.subtitle = element_text(size=20, face='bold', margin=margin(0,0,10,0), hjust=0.5),
      axis.text=element_text(size=rel(1.75), color='black'),
      axis.title=element_text(size=rel(1.95), color='black'),
      axis.title.y = element_text(margin=margin(0,10,0,0)),
      axis.title.x = element_text(margin=margin(10,0,0,0)),
      legend.position="right",
      panel.border=element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank()
    )
}

# Scatter plot function
scatter_plot <- function(data, xlabel="", ylabel="", title="", subtitle="", lm=FALSE, loess = FALSE) {
  p <- ggplot(data, aes_string(x=names(data)[1], y=names(data)[2], col=names(data)[3])) +
    labs(x=xlabel, y=ylabel, title=title, subtitle=subtitle) +
    geom_point(shape = 16, size = 2.5, show.legend = FALSE, alpha = .55) + 
    scatter_theme() + 
    scale_color_gradient(low = "#0091ff", high = "#f0650e")
  if (lm){
    if (loess){ p <- p + geom_smooth() }#+ scale_x_continuous(trans = "log1p")
    else{ p <- p + geom_smooth(method='lm') }
  }
}

# Define ggplot2 boxplot theme for boxplots
boxplot_theme <- function() {
  p <- theme(
    plot.title = element_text(size=30, hjust=0.5, margin=margin(0,0,20,0)),
    axis.title.y = element_text(colour="black", size=20, vjust=1.5),
    axis.title.x = element_text(colour="black", size=20, vjust=1.5, margin=margin(15,0,0,0)),
    axis.text.x = element_text(colour="black",size=rel(1.6)),
    axis.text.y = element_text(colour="black",size=rel(1.6)),
    axis.line = element_line(colour="black", size=rel(0.7)),
    axis.ticks.x = element_line(colour="black", size=rel(0.8)),
    axis.ticks.y = element_blank(),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    legend.position="none",
    legend.text=element_text(size=15),
    legend.title=element_blank(),
    legend.background=element_blank(),
    panel.border = element_blank()
  )
}
```

<!-- # Parse and filter data -->
```{r echo=FALSE, include=FALSE}
# Data
io   <- list()
io$base_dir   <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data-eb"
# io$base_dir <- "/Users/ricard/data/NMT-seq"
# io$base_dir <- "/Users/kapouranis/Documents/Andreas/Code/datasets/nmt-seq-data"
io$out_dir    <- "/home/cakapourani/Documents/Code/nmt-seq/rebuttal/EB/acc_conservation/out/lr/"
io$rna_file   <- paste0(io$base_dir, "/rna/parsed/sceset.rds")
io$met_file   <- paste0(io$base_dir, "/met/parsed/met_data.tsv.gz")
io$acc_file   <- paste0(io$base_dir, "/acc/parsed/profiles/cons_prof_prom_200_200_basis15_GpCcov10cellcov0.6.rds")
io$genelist_dir        <- paste0(io$base_dir, "/metadata")
io$plur_genes_file     <- paste0(io$genelist_dir, "/pluripotency_extended.tsv")
io$h3k4me3_genes_file  <- paste0(io$genelist_dir, "/H3K4me3.bed")
io$h3k27me3_genes_file <- paste0(io$genelist_dir, "/H3K27me3.bed")

# Filtering parameters
opts <- list()
opts$annos       <- c("prom_200_200")
opts$min.weight  <- 5      # Minimum weight per feature (met coverage)
opts$min.s       <- 15     # minimum number of samples to do the correlation
opts$gene_window <- 1e3    # window length for the overlap between genes and features
opts$quan_seq    <- seq(0, 1, length.out = 8) # Discretization steps
opts$axis_labels <- paste0("C", seq(1,7))     # Discretization steps
```

<!-- # Load scRNA data -->
```{r load_rna, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
sceset       <- readRDS(file = io$rna_file)  # Load expression as SCESet object
# Create data.table
rna_dt       <- exprs(sceset) %>% t %>% as.data.table(keep.rownames = "sample") %>% melt(id.vars = "sample", value.name = "expr", variable.name = "gene")
# Compute summary statistics for each gene across cells
rna_all_st   <- rna_dt[, .(mean_expr = mean(expr), median_expr = median(expr), var_expr = var(expr), sd_expr = sd(expr)), by = gene]
# Extract biological variance, i.e. overdispersion
rna_all_st   <- rna_all_st[, overdispersion := fData(sceset)$bioVar]
# Extract gene coordinates metadata to do the overlap
rna_metadata <- fData(sceset) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% .[,c("chr", "start", "end", "gene", "ens_id")]  %>% .[, chr := as.factor(sub("chr", "", chr))] %>% setnames("ens_id", "id")

rna_all_st   <- merge(rna_metadata[, c("chr", "id", "gene")], rna_all_st, by = c("gene")) # Merge using all samples
rna_dt       <- merge(rna_metadata[, c("chr", "id", "gene")], rna_dt, by = c("gene"))     # Merge using only samples that pass filtering
rm(sceset)
```

<!-- # Load scBS data -->
```{r load_met, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
# Load data
met_dt     <- fread(sprintf("zcat < %s", io$met_file)) %>% as.data.table %>% .[anno %in% opts$annos] %>% .[, rate := rate + rtruncnorm(.N, a = 0, b = Inf, mean = 0, sd = 0.01 )]
met_dt     <- met_dt[weight >= opts$min.weight]  # Keep only covered regions
keep_sites <- met_dt[,.(n = .N), by = c("id")] %>% .[n >= opts$min.s] %>% .$id  ## Filter id-gene pairs with small number of samples
met_dt     <- met_dt[id %in% keep_sites]
met_all_st <- met_dt[, .(mean_met = mean(rate), median_met = median(rate), var_met = var(rate), sd_met = sd(rate)), by = id] # Compute summary statistics for each gene across cells
rm(met_dt, keep_sites)
```

<!-- # Load sc-NOMe-seq data -->
```{r load_acc, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
# Load accessibility data
accrna_dt  <- readRDS(io$acc_file) %>% .[cells > 10] %>% .[, lr := log2(sc_NLL/bulk_NLL)] %>% .[, lr_levels := cut(abs(lr), breaks = quantile(abs(lr), probs = opts$quan_seq), labels = opts$axis_labels, include.lowest = TRUE)] # %>% .[log2ratio > 0.5 & log2ratio < 2 & coverage > 700] 

accrna_all <- merge(rna_all_st, accrna_dt, by = c("id")) %>% setorder(-lr)  # Merge with expression data using all samples 
metacc_all <- merge(met_all_st, accrna_dt, by = c("id")) %>% setorder(-lr)  # Merge with methylation data using all samples 

# # Process to get statistics only for cells that pass filtering
# dt_N <- NROW(accrna_dt)
# rna_filt <- data.table(id=character(dt_N), gene=character(dt_N), mean_expr=numeric(dt_N), median_expr=numeric(dt_N), var_expr = numeric(dt_N), sd_expr = numeric(dt_N))
# # Iterate over each region
# for (i in 1:NROW(accrna_dt)){
#   # Keep only statistics for cells that pass filtering
#   res <- rna_dt[id==accrna_dt$id[i] & sample%in%accrna_dt$cell_names[[i]], .(mean_expr=mean(expr), median_expr=median(expr), var_expr=var(expr), sd_expr=sd(expr)), by=c("id","gene")]
#   if (NROW(res) > 0 ){ set(rna_filt, i, 1:6, res) }
# }
# rna_filt    <- rna_filt %>% .[id != ""]  # Remove genes that we had no metadata information
# accrna_filt <- merge(rna_filt, accrna_dt, by = c("id")) %>% setorder(-lr) # Merge data
```

# Accessibility conservation and expression relationship

## Using likelihood information
To quantify the conservation of accessibility patterns around promoters we use the likelihood information from the fitted profiles. We pseudo-bulk the single-cell data and learn the joint profile $p(D_{1}, D_{2}, ... | Model)$ and the fit of the negative log-likelihood ( __bulk_NLL__ ). We also learn a separate profiles for each single-cell $p(D_{1} | Model_{1}), p(D_{2} | Model_{2}), ...$ and sum the negative log-likelihoods of each profile ( __sc_NLL__ ). Then we compute the log2 ratio :

$$
 LR = log_2(scNLL / bulkNLL)
$$
When $LR = 0$ we have the highest accessibility conservation (i.e. each cell has exactly the same accessibility profiles as the other cells). The smaller the value of $LR$, where $LR \leq 0$, the lower the conservation. 

The region for the accesibility profiles is $\pm 200$ around TSS. We observe that accessibility conservation around promoter regions is positively correlated with mean expression levels.
```{r associate, echo=FALSE, warning=FALSE, message=FALSE}
# Acc conservation - Mean expression association continouls LR all cells
p <- scatter_plot(accrna_all[, c("mean_expr", "lr", "coverage")], xlabel = "Mean expression", ylabel = "log2(scNLL/bulkNLL)", title = "All cells", subtitle = paste0("Pearson's r = ", round(cor(accrna_all$lr, accrna_all$mean_expr), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_mean_expr_cont_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())

# Acc conservation - Mean expression association discretised LR all cells
p <- ggplot(accrna_all[,c("lr_levels","mean_expr")], aes(x=lr_levels, y=mean_expr)) +
  geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#00BFC4") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
  labs(title="All cells", x="Conservation", y="Mean expression") + boxplot_theme()
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_mean_expr_discr_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())

# # Acc conservation - Mean expression association continouls LR filtered cells
# p <- scatter_plot(accrna_filt[, c("mean_expr", "lr", "coverage")], xlabel = "Mean expression", ylabel = "log2(scNLL/bulkNLL)", title = "Filtered cells", subtitle = paste0("r = ", round(cor(accrna_filt$lr, accrna_filt$mean_expr), 3)), lm = TRUE)
# invisible(pdf(file = paste0(io$out_dir, "lr_acc_mean_expr_cont_filt.pdf"), width = 12, height = 9)); p; invisible(dev.off())
# 
# # Acc conservation - Mean expression association discretised LR filtered cells
# p <- ggplot(accrna_filt[,c("lr_levels","mean_expr")], aes(x=lr_levels, y=mean_expr)) +
#   geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#00BFC4") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
#   labs(title="Filtered cells", x="Conservation", y="Mean expression") + boxplot_theme()
# p
# invisible(pdf(file = paste0(io$out_dir, "lr_acc_mean_expr_discr_filt.pdf"), width = 12, height = 9)); p; invisible(dev.off())
```

However we there does not seem to be some significant correlation between accessibility conservation and variability in expression.
```{r acc_var_expr, echo=FALSE, warning=FALSE, message=FALSE}
# Acc conservation - Expression variance association continouls LR all cells
p <- scatter_plot(accrna_all[, c("var_expr", "lr", "mean_expr")], xlabel = "Expression variance", ylabel = "log2(scNLL/bulkNLL)", title = "All cells", subtitle = paste0("Pearson's r =", round(cor(accrna_all$lr, accrna_all$var_expr), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_var_expr_cont_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())

# # Acc conservation - Expression variance association continouls LR all cells
# p <- scatter_plot(accrna_all[, c("var_expr", "lr", "mean_expr")], xlabel = "Expression variance", ylabel = "log2(scNLL/bulkNLL)", title = "All cells", subtitle = paste0("Pearson's r =", round(cor(accrna_all$lr, accrna_all$var_expr), 3)), lm = TRUE)
# p
# invisible(pdf(file = paste0(io$out_dir, "lr_acc_var_expr_cont_filt.pdf"), width = 12, height = 9)); p; invisible(dev.off())
```

Also when trying to associate acc conservation with gene expression __overdispersion__ (i.e. biological variability) we find no correlation, indicating the variability on chromatin accessibility is not necessarily assocated to variability in gene expression.

```{r acc_over_expr, echo=FALSE, warning=FALSE, message=FALSE}
# Acc conservation - Expression overdispersion association continouls LR all genes
p <- scatter_plot(accrna_all[, c("overdispersion", "lr", "mean_expr")], xlabel = "Expression overdispersion", ylabel = "log2(scNLL/bulkNLL)", title = "All overdispersion", subtitle = paste0("Pearson's r =", round(cor(accrna_all$lr, accrna_all$overdispersion), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_overdisp_expr_cont_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())

# Acc conservation - Expression overdispersion association continouls LR positive dispersed genes
accrna_overdisp <- copy(accrna_all)
accrna_overdisp <- accrna_overdisp[overdispersion > 0]
p <- scatter_plot(accrna_overdisp[, c("overdispersion", "lr", "mean_expr")], xlabel = "Expression overdispersion", ylabel = "log2(scNLL/bulkNLL)", title = "Positive overdispersion", subtitle = paste0("Pearson's r =", round(cor(accrna_overdisp$lr, accrna_overdisp$overdispersion), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_overdisp_expr_cont_filt.pdf"), width = 12, height = 9)); p; invisible(dev.off())
```



## Using profile weight distance
We can quantify the conservation of accessibility patterns using the Euclidean distance between the profile weights of each cell. Essentially similar weights would imply similar profiles hence the distance would be lower for similar accessibility patterns across cells.

As we can observe this does not correlate well with expression levels, which might indicate that this metric is not the most appropriate or there is no correlation between the two layers.
```{r acc_mean_expr_2, echo=FALSE, warning=FALSE, message=FALSE}
# Weight distnance Acc conservation - Mean expression association continouls LR filtered cells
p <- scatter_plot(accrna_all[, c("mean_expr", "weight_dist", "coverage")], xlabel = "Mean expression", ylabel = "Weight distance", title = "All cells", subtitle = paste0("r = ", round(cor(accrna_all$weight_dist, accrna_all$mean_expr), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "wd_acc_mean_expr_cont_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())
```

Also we do not observe any relationship between the variability of expression.
```{r acc_var_expr_2, echo=FALSE, warning=FALSE, message=FALSE}
# Weight distnance Acc conservation - Expression variance association continouls LR filtered cells
p <- scatter_plot(accrna_all[, c("var_expr", "weight_dist", "mean_expr")], xlabel = "Expression variance", ylabel = "Weight distance", title = "All cells", subtitle = paste0("Pearson's r =", round(cor(accrna_all$weight_dist, accrna_all$var_expr), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "wd_acc_var_expr_cont_all.pdf"), width = 12, height = 9)); p; invisible(dev.off())
```

## Using weights from smaller windows
TODO... Idea is to split the big window in smaller windows and evaluate the accessibility on the small windows obtaining a vector of accesibility patterns across the whole region. Then compute the Euclidean distance between all the cells to obtain a metric of similarity (Maybe better that using the weights of the basis functions). 

## GpC coverage as a confounding factor
Interestingly, we observe that the weight distance metric is highly correlated with the GpC coverage, something that we do not observe for the LR metric, indicating that the weight distance would not be the more appropriate metric for understanding the relationship between expression and accessibility.
```{r acc_var_sssexpr_2, echo=FALSE, warning=FALSE, message=FALSE, fig.height=8, fig.width=17}
# CpG coverage vs Weight distance metric association
d1 <- scatter_plot(accrna_all[, c("coverage", "weight_dist", "mean_expr")], xlabel = "Coverage", ylabel = "Weight distance", title = "", lm = TRUE)
# CpG coverage vs LR ratio metric association
d2 <- scatter_plot(accrna_all[, c("coverage", "lr", "mean_expr")], xlabel = "Coverage", ylabel = "log2(scNLL/bulkNLL)", title = "", lm = TRUE)
# Create the two plots together
p <- plot_grid(d1, d2, labels = c("", ""), label_size = 20, ncol = 2, nrow = 1); p
# Store them to pdf
invisible(pdf(file = paste0(io$out_dir, "cov_wd_lr_assoc.pdf"), width = 16, height = 9)); print(p); invisible(dev.off())
```

# Mean variance relationship of expression data
```{r mean_var_expr, echo=FALSE, warning=FALSE, message=FALSE, eval=TRUE}
p <- scatter_plot(accrna_all[, c("mean_expr", "var_expr", "lr")], xlabel = "Mean expression", ylabel = "Expression variance", title = "", lm = TRUE, loess = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "expr_mean_var_assoc.pdf"), width = 16, height = 9)); print(p); invisible(dev.off())
```


# Association with methylation
The region for the accesibility profiles is $\pm 200$ around TSS. We observe that accessibility conservation around promoter regions is not correlated with methylation rate and methylation variance.
```{r associate_met, echo=FALSE, warning=FALSE, message=FALSE}
# Accessibility conservation - Mean methylation association continuous LR all cells
p <- scatter_plot(metacc_all[, c("mean_met", "lr", "coverage")], xlabel = "Mean methylation", ylabel = "log2(scNLL/bulkNLL)", title = "All cells", subtitle = paste0("Pearson's r = ", round(cor(metacc_all$lr, metacc_all$mean_met), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_met_rate_cont_all.pdf"), width = 16, height = 9)); p; invisible(dev.off())

# Accessibility conservation - Methylation variance association continuous LR all cells
p <- scatter_plot(metacc_all[, c("sd_met", "lr", "coverage")], xlabel = "SD methylation", ylabel = "log2(scNLL/bulkNLL)", title = "All cells", subtitle = paste0("Pearson's r = ", round(cor(metacc_all$lr, metacc_all$sd_met), 3)), lm = TRUE)
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_met_sd_cont_all.pdf"), width = 16, height = 9)); p; invisible(dev.off())

# Accessibility conservation - Mean methylation association discretised LR all cells
p <- ggplot(metacc_all[,c("lr_levels","mean_met")], aes(x=lr_levels, y=mean_met)) +
  geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#00BFC4") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
  scale_y_continuous(limits = c(0, 15)) + labs(title=NULL, x="Conservation", y="Mean methylation") + boxplot_theme()
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_met_rate_discr_all.pdf"), width = 16, height = 9)); p; invisible(dev.off())

# Accessibility conservation - Methylation variance association discretised LR all cells
p <- ggplot(metacc_all[,c("lr_levels","sd_met")], aes(x=lr_levels, y=sd_met)) +
  geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#00BFC4") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
  scale_y_continuous(limits = c(0, 20)) + labs(title=NULL, x="Conservation", y="SD methylation") + boxplot_theme()
p
invisible(pdf(file = paste0(io$out_dir, "lr_acc_met_sd_discr_all.pdf"), width = 16, height = 9)); p; invisible(dev.off())
```

## GO analysis
Here we perform Gene Ontology analysis to see whether specific clusters of genes are enriched for specific pathways. 
```{r go_enr_list, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
io$reactomeFile  <- paste0(io$genelist_dir, "/mouse_v75_reactome.rds")
io$reactomeFile2 <- paste0(io$genelist_dir, "/AllPathways.txt")

# Read reactome file and change gene names to Upper case
reactome           <- readRDS(io$reactomeFile); colnames(reactome) <- toupper(colnames(reactome))
# Read pathways file to change from ID to description pathway
tmp                <- read.table(io$reactomeFile2, header=F, quote="", sep="\t", stringsAsFactors=F)[,c(1,2)]
reactome_meta      <- tmp[,2]; names(reactome_meta) <- tmp[,1]
rownames(reactome) <- stringr::str_replace_all(rownames(reactome), reactome_meta)
feature.sets       <- reactome  # Store results to the following object

opts$min.size      <- 10  # Filter feature sets with small number of features
feature.sets       <- feature.sets[rowSums(feature.sets)>=opts$min.size, ]
  
# Remove features that do not intersect between the feature sets and the data
accrna_tmp        <- copy(accrna_all)
features          <- intersect(toupper(accrna_tmp$gene), colnames(feature.sets))
foo_filt          <- accrna_tmp[,gene:=toupper(gene)] %>% .[gene %in% features]
feature.sets_filt <- feature.sets[,features]

# Melt and merge data to do the fisher's test
asd <- melt.data.table(as.data.table(feature.sets_filt, keep.rownames="pathway"), id.vars = "pathway", variable.name="gene", value.name="value") %>% merge(foo_filt[,c("gene","lr_levels")], allow.cartesian=T, by="gene")
# Get GO list for each cluster
go_list <- list()
for (c in unique(sort(asd$lr_levels))) {
  go_list[[c]] <- asd[,.(p=fisher.test(x=matrix(data=c(.SD[lr_levels==c & value==1,.N], .SD[lr_levels!=c & value==1,.N],
                                                       .SD[lr_levels==c & value==0,.N], .SD[lr_levels!=c & value==0,.N]), 
                                                nrow=2, ncol=2), alternative="greater")[["p.value"]]), by="pathway"] %>% 
    .[,padj:=p.adjust(p, method = "fdr")] %>% setorder(p)
  
  # Create a nice formatted table
  print(knitr::kable(go_list[[c]][1:30], caption = paste0("GO enrichment: Cluster ", c)))
}

saveRDS(go_list, file = paste0(io$out_dir, "GO_list_lr.rds"))
```

## Histone marks
Here we perform Histone mark enrichment analysis to see whether specific clusters of genes are enriched for specific marks. We also check the proportion of H3K27m3 vs H3k4m3, and we observe that the more variable is the accessibility the higher the proportion of the repressive H3K27m3 mark is. 
```{r plur_genelist, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
# Read histone marks files
h3k4_genes_file <- fread(io$h3k4me3_genes_file) %>% as.data.table %>% .[, c("V1", "V2", "V3")] %>% setnames(c("chr", "start", "end")) %>% setkey(chr, start, end) %>% .[, chr := as.factor(sub("chr", "", chr))] %>% .[, id := paste0("hek4m3_", 1:.N)]
h3k27_genes_file <- fread(io$h3k27me3_genes_file) %>% as.data.table %>% .[, c("V1", "V2", "V3")] %>% setnames(c("chr", "start", "end")) %>% setkey(chr, start, end) %>% .[, chr := as.factor(sub("chr", "", chr))] %>% .[, id := paste0("h3k27me3_", 1:.N)]
# Get RNA metdata to do the overlapping
rna_metadata_tmp <- rna_metadata[, c("chr", "start", "end", "gene")] %>%
  .[,c("start", "end") := list(start - opts$gene_window, start + opts$gene_window)] %>% setkey(chr, start, end)
# Do the overlap
ov_h3k4  <- foverlaps(h3k4_genes_file, rna_metadata_tmp, nomatch = 0) %>% .[,c("gene")] %>% unique
ov_h3k27 <- foverlaps(h3k27_genes_file, rna_metadata_tmp, nomatch = 0) %>% .[,c("gene")] %>% unique
ov_h3    <- Reduce(intersect, list(ov_h3k4$gene, ov_h3k27$gene)) %>% as.data.table %>% setnames(c("gene"))

# Keep only genes related to enahncers to get relative enrichment
accrna_enr  <- accrna_all[gene %in% ov_h3k4$gene | gene %in% ov_h3k27$gene]
N_h3k4   <- NROW(accrna_enr[gene %in% ov_h3k4$gene])
N_h3k27  <- NROW(accrna_enr[gene %in% ov_h3k27$gene])
cls <- length(unique(accrna_all$lr_levels))   # Total number of clusters k
h3k4 = h3k27 = h3 <- vector(mode = "list", length = cls)
h3k4_c = h3k27_c <- vector(mode = "numeric", length = cls)
for (i in 1:cls){  # Iterate over each cluster and extract number of genes related to each histone mark
  h3k4[[i]] <- accrna_enr[gene %in% ov_h3k4$gene & lr_levels == paste0("C",i)]
  h3k27[[i]] <- accrna_enr[gene %in% ov_h3k27$gene & lr_levels == paste0("C",i)]
  h3k4_c[i] <- NROW(h3k4[[i]])
  h3k27_c[i] <- NROW(h3k27[[i]])
}
# Create data frame with the summary results
df <- data.frame(rbind(as.integer(h3k4_c), as.integer(h3k27_c), round(h3k27_c/h3k4_c, 3)))
colnames(df) <- paste0("Cluster", seq(1:cls))
rownames(df) <- c("H3K4me3", "H3K27me3", "Ratio")
# Create a nice formatted table
knitr::kable(df, caption = "Histone Marks enrichment")
  
# for (i in 1:cls){
#   hits <- Reduce(intersect, list(h3k4[[i]]$id , h3k27[[i]]$id)) 
#   h3[i] <- accrna_enr[gene %in% ov_h3k4me3$gene & gene %in% ov_h3k27me3$gene & clusters == i])
# }
#plur_genes_file <- fread(io$plur_genes_file) %>% as.data.table %>% setnames(c("id", "gene"))
# plur_mrg_dt <- merge(plur_genes_file, accrna_all_factor, by = "id")
```