---
title: "NMT-seq EB: general statistics of genomic features"
output: 
  BiocStyle::html_document: 
    toc: false
    fig_width: 7
    fig_height: 7
---

```{r load_modules, echo=FALSE, include=FALSE}
library(weights)
library(data.table)
library(purrr)
library(ggplot2)
```

```{r, echo=FALSE}

boxplot_theme <- function() {
  p <- theme(
    plot.title = element_blank(),
    axis.title.y = element_blank(),
    axis.title.x = element_text(colour="black", size=20, vjust=1.5, margin=margin(10,0,0,0)),
    axis.text.x = element_text(colour="black",size=rel(1.7)),
    axis.text.y = element_text(colour="black",size=rel(1.7)),
    axis.line = element_line(colour="black", size=rel(0.7)),
    axis.ticks.x = element_line(colour="black", size=rel(1.0)),
    axis.ticks.y = element_line(colour="black", size=rel(1.0)),
    # axis.ticks.y = element_blank(),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    legend.position="top",
    legend.text=element_text(size=15),
    legend.title=element_blank(),
    legend.background=element_blank(),
    panel.border = element_blank()
  )
}
```

```{r define_options, echo=FALSE}

## I/O ##
io <- list()
io$met_file  <- "/Users/ricard/data/NMT-seq_EB/met/parsed/met_data.tsv.gz"
io$acc_file  <- "/Users/ricard/data/NMT-seq_EB/acc/parsed/acc_data.tsv.gz"
io$sample.metadata  <- "/Users/ricard/data/NMT-seq_EB/sample_sheet.csv"
io$outdir  <- "/Users/ricard/NMT-seq/rebuttal/EB/stats/out"

## Options ##
opts <- list()

# Define which cells to use
opts$cells <- "all"
if (opts$cells == "all") {
    opts$acc_cells <- fread(io$sample.metadata) %>% .[pass_accQC==T,sample]
    opts$met_cells <- fread(io$sample.metadata) %>% .[pass_metQC==T,sample]
}

# Define which annotations to use and the order for plotting
opts$acc_annos <- c("prom_2000_2000_cgi","prom_2000_2000_noncgi","active_enhancers","super_enhancers","primed_enhancers","Nanog","Oct4","DHS","p300","exons","introns")
opts$met_annos <- c("prom_2000_2000_cgi","prom_2000_2000_noncgi","active_enhancers","super_enhancers","primed_enhancers","p300","Oct4","Nanog","DHS","exons","introns")

# Filter features with low coverage
# opts$min.weight <- 1
# opts$min.coverage <- 0.5

opts$anno_names <- c(
  "prom_50_50_cgi"="CGI promoters",
  "prom_50_50_noncgi"="non-CGI promoters",
  "genebody"="Gene body",
  "exons"="Exons",
  "introns"="Introns",
  "active_enhancers"="Active enhancers",
  "super_enhancers"="Super enhancers",
  "primed_enhancers"="Primed enhancers",
  "prom_2000_2000_cgi"="CGI promoters",
  "prom_2000_2000_noncgi"="non-CGI promoters"
)

opts$anno_order <- c("CGI promoters","non-CGI promoters", "Exons", "Introns", "Active enhancers", "Super enhancers", "Primed enhancers", "DHS", "p300", "Nanog", "Oct4")

```

```{r load_metacc, echo=FALSE, include=FALSE}
met <- fread(sprintf("zcat < %s",io$met_file), showProgress=F) %>% .[anno %in% opts$met_annos] %>% .[,anno:=factor(anno, levels=opts$met_annos)]%>% .[,context:="CG"]
acc <- fread(sprintf("zcat < %s",io$acc_file), showProgress=F) %>% .[anno %in% opts$acc_annos]  %>% .[,anno:=factor(anno, levels=opts$acc_annos)]%>% .[,context:="GC"]
metacc <- rbind(met,acc)

# Merge with metadata
metacc <- merge(fread(io$sample.metadata)[,c("sample","lineage")], metacc, by="sample")
```


```{r filter_data, echo=FALSE}

# Remove features with low weight
# metacc <- metacc[weight >= opts$min.weight]

# Remove features with low coverage
# metacc <- metacc[, cov:=.N/length(opts$cells), by=c("id","anno")] %>% .[cov >= opts$min.coverage] %>% .[,cov:=NULL]

```


```{r calculate_stats, echo=FALSE, include=FALSE}
feature_stats <- metacc[,.(mean=mean(rate, na.rm=T), var=var(rate, na.rm=T)), by=c("anno","id","context","lineage")] %>%
  .[,anno:=stringr::str_replace_all(anno, opts$anno_names)] %>%
  .[,anno:=factor(anno,levels = opts$anno_order)] %>% droplevels

# feature_stats <- metacc[,.(mean=mean(rate, na.rm=T), wtd_mean=weighted.mean(rate, weight, na.rm=T),
#                var=var(rate, na.rm=T), wtd_var=wtd.var(rate, weight, na.rm=T),
#                weight=sum(weight) ), by=c("anno","id","context")]


```

<!-- Mean methylation/accessibility rate per genomic context and lineage -->
```{r mean, echo=FALSE, include=TRUE}
p1 <- ggplot(feature_stats[context=="CG",], aes(x=anno, y=mean)) +
  geom_boxplot(aes(fill=lineage), alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Mean methylation rate") +
  scale_fill_manual(values=c("orangered4","orangered1")) +
  coord_flip() +
  boxplot_theme()
p1

p2 <- ggplot(feature_stats[context=="GC",], aes(x=anno, y=mean)) +
  geom_boxplot(aes(fill=lineage), alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Mean accessibility rate") +
  scale_fill_manual(values=c("dodgerblue4","dodgerblue1")) +
  coord_flip() +
  boxplot_theme()
p2

# pdf("/Users/ricard/NMT-seq/rebuttal/EB/stats/out/feature_met_means.pdf")
# print(p1)
# dev.off()

# pdf("/Users/ricard/NMT-seq/rebuttal/EB/stats/out/feature_acc_means.pdf")
# print(p2)
# dev.off()

```


<!-- Variance across cells on methylation/accessibility rate per genomic context -->
(Q) Pluripotent and differentiated lineages have differnet number of cells, is it fair to compare the variance?
```{r met_var, echo=FALSE, include=TRUE, warning=FALSE}

# p <- ggplot(feature_stats[context=="GC"], aes(x=anno, y=var)) +
#   geom_boxplot(alpha=0.8, coef=0, outlier.shape=NA, fill="#00BFC4") +
#   ggtitle("") + xlab("") + ylab("Cell-to-cell variance on the accessibility rate") +
#   coord_flip(ylim=c(0,2000)) +
#   boxplot_theme()
# print(p)

p3 <- ggplot(feature_stats[context=="CG"], aes(x=anno, y=var)) +
  geom_boxplot(aes(fill=lineage), alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Cell-to-cell variance on the methylation rate") +
  scale_fill_manual(values=c("orangered4","orangered1")) +
  coord_flip(ylim=c(0,2500)) +
  boxplot_theme()
p3

p4 <- ggplot(feature_stats[context=="GC"], aes(x=anno, y=var)) +
  geom_boxplot(aes(fill=lineage), alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Cell-to-cell variance on the methylation rate") +
  scale_fill_manual(values=c("dodgerblue4","dodgerblue1")) +
  coord_flip(ylim=c(0,2100)) +
  boxplot_theme()
p4


# pdf("/Users/ricard/NMT-seq/rebuttal/EB/stats/out/feature_met_var.pdf")
# print(p3)
# dev.off()
# 
# pdf("/Users/ricard/NMT-seq/rebuttal/EB/stats/out/feature_acc_var.pdf")
# print(p4)
# dev.off()
```


```{r}
pdf(paste0(io$outdir,"/boxplots.pdf"), height=12, width=17)
print(cowplot::plot_grid(p1,p3,p2,p4, labels = c("a","b","c","d"), label_size=20, ncol=2, nrow=2))
dev.off()
```
