---
title: "NMT-Seq EB: coupling along pseudotime"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(purrr))
source("/Users/ricard/NMT-seq_EB/pseudotime/utils.R")
```

```{r}
```

```{r}
opts <- list()
opts$metrna_annos <- c("Active enhancers","Primed enhancers","Super enhancers","CTCF","Nanog","non-CGI promoters","Oct4","p300")
opts$accrna_annos <- c("Active enhancers","Primed enhancers","Super enhancers","CTCF","Nanog","non-CGI promoters","Oct4","p300")
```

<!-- Load correlation results -->
```{r}

# Load methylation/RNA correlations
metrna <- fread("/Users/ricard/NMT-seq_EB/correlations/out/metrna/second_submission/metrna_cor_features.tsv", header=T) %>% .[,type:="CG"]# %>% .[anno%in%opts$metrna_annos]

# Load accessibility/RNA correlations
accrna <- fread("/Users/ricard/NMT-seq_EB/correlations/out/accrna/second_submission/accrna_cor_features.tsv", header=T) %>% .[,type:="GC"]# %>% .[anno%in%opts$accrna_annos]

# Load accessibility/methylation correlations
metacc <- fread("/Users/ricard/NMT-seq_EB/correlations/out/metacc/second_submission/metacc_cor_features.tsv", header=T) %>% .[,type:="CGC"]# %>% .[anno%in%opts$metacc_annos]

# Load pseudotime trajectory
pseudotime <- fread("/Users/ricard/NMT-seq_EB/pseudotime/out/destiny.tsv", header=T )

```


```{r}
tmp <- rbind(metrna,accrna,metacc) %>% merge(pseudotime, by="sample") %>%
  .[,pseudotime:=pseudotime]
```


<!-- Plot all genomic contexts for methylation/rna -->
```{r}
p <- ggplot(tmp[type=="CG"], aes(x=pseudotime, y=r, color=anno)) +
  stat_smooth(aes(fill=anno), method="lm", se=T, alpha=0.1) +
  labs(x="Pseudotime", y="Methylation/RNA correlation coefficient", color="") +
  guides(fill=FALSE) +
  theme_bw() + theme_pub()
print(p)
```


<!-- Plot all genomic contexts for accessibility/rna -->
```{r}
p <- ggplot(tmp[type=="GC"], aes(x=pseudotime, y=r, color=anno)) +
  # geom_point() +
  stat_smooth(method="lm", se=F) +
  labs(x="Pseudotime", y="Correlation coefficient", color="") +
  theme_bw() + theme_pub()
print(p)
```

<!-- Plot all genomic contexts for accessibility/methylation -->
```{r}

# opts$metacc_annos <- c("Active enhancers","Primed enhancers","Super enhancers","CTCF","Nanog","non-CGI promoters","Oct4","p300")
opts$metacc_annos <- c("Super enhancers","CTCF","Nanog","non-CGI promoters","Oct4","p300")
# opts$metacc_annos <- c("Super enhancers")

p <- ggplot(tmp[type=="CGC" & anno%in%opts$metacc_annos], aes(x=pseudotime, y=r, color=anno)) +
  # geom_point(size=1, data=tmp[type=="CGC" & anno%in%c("p300")]) +
  # geom_point(size=1, alpha=0.5) +
  # stat_smooth(method="lm", se=F) +
  stat_smooth(aes(fill=anno), method="lm", se=T, alpha=0.1) +
  # stat_smooth(method="loess", se=F) +
  labs(x="Pseudotime", y="Methylation/Accessibility correlation coefficient", color="") +
  guides(fill=FALSE) +
  # scale_x_continuous(breaks = sort(unique(tmp$pseudotime)), 
                   # limits = c(-0.25,0.22), expand = c(0,0)) +
  theme_bw() + theme_pub()
print(p)

# pdf("/Users/ricard/NMT-seq/rebuttal/EB/pseudotime/out/metacc_correlation.pdf", width = 8, height = 6)
# print(p)
# dev.off()
```


<!-- Plot individual genomic contexts for methylation/RNA -->
```{r}

# opts$metrna_annos <- c("Exons","CTCF","Nanog","non-CGI promoters","Oct4","p300")

annos <- unique(metrna$anno)
colors <- sample_colors(length(annos))
names(colors) <- annos

for (a in annos) {
  p <- ggplot(tmp[type=="CG" & anno==a], aes(x=pseudotime, y=r, color=anno)) +
    ggtitle(a) +
    geom_point(size=2.5, alpha=0.75, color=colors[a]) +
    stat_smooth(method="lm", se=T, alpha=0.1, color=colors[a], fill=colors[a], size=1.5) +
    labs(x="Pseudotime (Diffusion Component 1)", y="Met/RNA correlation coefficient", color="") +
    # coord_cartesian(ylim=c(-0.41,0.0)) +
    guides(fill=FALSE) +
    theme_bw() + theme_pub()
  # pdf(paste0("/Users/ricard/NMT-seq/rebuttal/EB/pseudotime/out/metacc_",a,".pdf"), width = 3, height = 4, useDingbats = F)
  print(p)
  # dev.off()
}

```

<!-- Plot individual genomic contexts for accessibility/methylation -->
```{r}

opts$metacc_annos <- c("Exons","CTCF","Nanog","non-CGI promoters","Oct4","p300")

colors <- sample_colors(length(opts$metacc_annos))
names(colors) <- opts$metacc_annos

for (a in opts$metacc_annos) {
  p <- ggplot(tmp[type=="CGC" & anno==a], aes(x=pseudotime, y=r, color=anno)) +
    geom_point(size=2.5, alpha=0.75, color=colors[a]) +
    stat_smooth(method="lm", se=T, alpha=0.1, color=colors[a], fill=colors[a], size=1.5) +
    labs(x="Pseudotime (Diffusion Component 1)", y="Methylation/Accessibility correlation coefficient", color="") +
    coord_cartesian(ylim=c(-0.41,0.0)) +
    guides(fill=FALSE) +
    theme_bw() + theme_pub()
  # pdf(paste0("/Users/ricard/NMT-seq/rebuttal/EB/pseudotime/out/metacc_",a,".pdf"), width = 3, height = 4, useDingbats = F)
  print(p)
  # dev.off()
}

```

<!-- Scatter plots -->
```{r}

```