---
title: "NMT-seq: preprocessing of expression data"
author: "Ricard Argelaguet"
output: 
  BiocStyle::html_document: 
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(scater)
library(scran)
library(ggplot2)
source("/Users/ricard/NMT-seq/Rutils/stats_utils.R")
```

```{r fncs, echo=FALSE}
```


```{r define_opts, echo=FALSE, include=FALSE}

## I/O ##
io <- list()
io$in.gene_metadata <- "/Users/ricard/data/ensembl/mouse/v87/BioMart/mRNA/Mmusculus_genes_BioMart.87.txt"
io$in.sample_metadata <- "/Users/ricard/data/NMT-seq_EB/sample_sheet.csv"
io$in.raw_counts <- "/Users/ricard/data/NMT-seq_EB/rna/parsed/raw_counts.txt"
io$out.file <- "/Users/ricard/data/NMT-seq_EB/rna/parsed/sceset.rds"
io$plots.outdir <- "/Users/ricard/NMT-seq/rebuttal/EB/preprocessing/out"

## Options ##
opts <- list()

# opts$coverage_threshold <- 3e5    # Minimum library size (coverage)
opts$coverage_threshold <- 1e5      # Minimum library size (coverage)
opts$features_threshold <- 3000     # Minimum number of expressed features
opts$top50_threshold <- 25          # Maximum fraction of reads accounting for the top 50 features
opts$MT_threshold <- 0.10           # Maximum fraction of reads mapping to mithocondrial genes

```

```{r load_data, echo=TRUE, include=FALSE}
raw.counts <- fread(io$in.raw_counts)
count.mat <- raw.counts[,-c("ens_id")] %>% as.matrix()
row.names(count.mat) = raw.counts$ens_id
count.mat <- count.mat[,sort(colnames(count.mat))]
```


```{r load_gene_metadata, echo=FALSE, include=FALSE}
# gene_metadata <- read.csv(io$in.gene_metadata, sep="\t", stringsAsFactors=FALSE, quote="", header=T)
gene_metadata <- fread(io$in.gene_metadata, sep="\t", header=T)

# Include ERCC to feature metadata
# ercc <- raw.counts[Chromosome == "ERCC"]$Gene
# gene_metadata <- rbind(gene_metadata, data.table(symbol=ercc, chr="chrERCC"), fill=TRUE)

# Define mithocondrial genes
mt <- gene_metadata$symbol[gene_metadata$chr == "chrMT"]

# remove duplicated genes (THIS SHOULD BE DONE WHEN CREATING THE ANNOTATION)
gene_metadata <- gene_metadata[!duplicated(gene_metadata$symbol),]

# remove uninteresting genes 
genes <- rownames(count.mat)[rownames(count.mat) %in% gene_metadata$ens_id]
genes_not <- gene_metadata$ens_id[!gene_metadata$ens_id %in% rownames(count.mat)]
count.mat <- count.mat[genes,]
gene_metadata <- gene_metadata[ens_id %in% genes,]

# convert from ens_id to gene symbol in count matrix, sort and match rows with metadata
# gene_metadata <- gene_metadata %>% tibble::remove_rownames() %>% tibble::column_to_rownames("symbol")
rownames(count.mat) <- gene_metadata %>% setkey(ens_id) %>% .[rownames(count.mat),symbol] 
gene_metadata %>% setnames("symbol","gene") %>% setkey(gene)
count.mat <- count.mat[gene_metadata$gene,]
stopifnot(all(rownames(count.mat) == gene_metadata$symbol))
```

```{r load_sample_metadata, echo=FALSE, include=FALSE}
sample_metadata <- fread(io$in.sample_metadata, header=T) %>% setkey(sample)
count.mat <- count.mat[,sample_metadata$sample]
```

```{r create_sceset, echo=FALSE, include=FALSE}

# Create featureData
fdata <- gene_metadata %>% as.data.frame %>% tibble::column_to_rownames("gene") %>% new(Class = "AnnotatedDataFrame")

# Create phenoData
pdata <- sample_metadata %>% as.data.frame %>% tibble::column_to_rownames("sample") %>% new(Class = "AnnotatedDataFrame")

# create SCEset object
sce = newSCESet(countData = count.mat, phenoData = pdata, featureData = fdata)

# Calculate quality metrics
sce = calculateQCMetrics(sce, feature_controls=list(Mt=rownames(sce) %in% mt))
```

```{r filter_samples, echo=FALSE, include=TRUE}

# Library size
libsize.drop <- sce$total_counts < opts$coverage_threshold
foo <- data.table(sample=colnames(sce), size=sce$total_counts, color=c("black","red")[as.numeric(libsize.drop)+1]) %>% 
  setkey(size) %>% .[,col:=size] %>% .[,sample:=factor(sample,levels=sample)]

p1 <- ggplot(foo, aes(x=sample, y=size)) +
  # geom_bar(aes(fill=col), stat='identity', position="dodge", color="black") +
  geom_bar(stat='identity', position="dodge", color="black", fill="#3BAA3C") +
  geom_hline(yintercept=opts$coverage_threshold, colour="black", linetype="dashed") +
  # scale_fill_gradient(low="red", high="green") +
  labs(y="Library size") +
  barplot_theme() +
  theme(
    legend.position = "none",
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=rel(1.8)),
    axis.text.x = element_text(colour="black", color=foo$color, angle=90, size=10, vjust=0.5, hjust=1.0)
  )
print(p1)

# pdf(file="/Users/ricard/NMT-seq/rebuttal/2i_vs_serum/QC/out/rna_library_size.pdf", width=10, height=7)
# print(p1)
# dev.off()

# Number of expressed genes
feature.drop <- sce$total_features < opts$features_threshold
foo <- data.table(sample=colnames(sce), features=sce$total_features, color=c("black","red")[as.numeric(feature.drop)+1]) %>%
  setkey(features) %>% .[,col:=features] %>% .[,sample:=factor(sample,levels=sample)]
p2 <- ggplot(foo, aes(x=sample, y=features)) +
  # geom_bar(aes(fill=col), stat='identity', position="dodge", color="black") +
  geom_bar(stat='identity', position="dodge", color="black", fill="#3BAA3C") +
  geom_hline(yintercept=opts$features_threshold, colour="black", linetype="dashed") +
  # scale_fill_gradient(low="red", high="green") +
  labs(y="Total number of expressed genes") +
  barplot_theme() +
  theme(
    legend.position = "none",
    axis.title.x = element_blank(),
    axis.title.y = element_text(size=rel(1.8)),
    axis.text.x = element_text(colour="black", color=foo$color, angle=90, size=10, vjust=0.5, hjust=1.0)
  )
print(p2)

# pdf(file="/Users/ricard/NMT-seq/rebuttal/2i_vs_serum/QC/out/rna_nfeatures.pdf", width=10, height=7)
# print(p2)
# dev.off()

# Proportion of reads accounting for the top 50 features
# top50.drop <- sce$pct_counts_top_50_features > opts$top50_threshold
# foo <- data.table(sample=colnames(sce), pct=sce$pct_counts_top_50_features) %>% setkey(pct) %>% .[,col:=pct] %>% .[,sample:=factor(sample,levels=sample)]
# p <- barPlot(foo, ylabel="Fraction of reads accounting for the top 50 features") + 
#   geom_hline(yintercept=opts$top50_threshold, colour="blue", linetype="dashed") +
#   scale_fill_gradient(low="red", high="green") +
#   theme_barplot_pub()
# print(p)

# Samples with large proportion of mithocondrial genes
# mt.drop <- sce$pct_counts_feature_controls_Mt > opts$MT_threshold*100
# foo <- data.table(
#   sample=colnames(sce),
#   mt=sce$pct_counts_feature_controls_Mt,
#   color=c("black","red")[as.numeric(mt.drop)+1]
# ) %>% setkey(mt) %>% .[,col:=mt] %>% .[,sample:=factor(sample,levels=sample)]
# 
# p3 <- ggplot(foo, aes(x=sample, y=mt, fill=col)) +
#   geom_bar(stat='identity', position="dodge") +
#   geom_hline(yintercept=opts$MT_threshold*100, colour="blue", linetype="dashed") +
#   scale_fill_gradient(low="red", high="green") +
#   labs(y="Fraction of reads mapping to mithocondrial genes") +
#   barplot_theme() +
#   theme(
#     legend.position = "none",
#     axis.text.x = element_text(colour="black", color=foo$color, angle=90, size=10, vjust=0.5, hjust=1.0)
#   )
# print(p3)

p <- cowplot::plot_grid(p1, p2, ncol=2, nrow=1)

pdf(file=paste0(io$plots.outdir,"/qc_expr.pdf"), width=19, height=7)
print(p)
dev.off()
```



```{r remove_outliers, echo=FALSE, include=TRUE}

# Remove outliers
# drop.samples <- colnames(sce)[( libsize.drop | feature.drop | mt.drop )]
# drop.samples <- colnames(sce)[( libsize.drop | feature.drop )]
print(drop.samples)

# sample_metadata[,pass_rnaQC:=ifelse(sample%in%drop.samples,FALSE,TRUE)]
# fwrite(sample_metadata,io$in.sample_metadata, sep=",", col.names = T, row.names = F)

sce <- sce[,!colnames(sce) %in% drop.samples]
sce <- calculateQCMetrics(sce)
```

<!-- # Filter genes -->
<!-- We don't filter genes here, they are useful for some further analyses. -->
<!-- Lowly expressed genes (mean count <1) can be removed from the RNAseq. This leaves the following number of genes for analysis: -->
<!-- ```{r include=TRUE} -->
<!-- sce.all.genes = sce # also keep a dataset with all genes -->
<!-- keep = rowMeans(counts(sce)) >= 1 -->
<!-- sce = sce[keep,]  -->
<!-- sum(keep) -->
<!-- ``` -->

# Normalisation and log transformation
Transcript counts are now normalised based on size factors using the convolution approach from the scran package.
Lowly expressed genes are removed before normalisation but they are included afterwards, since they are interesting for some analysis.
```{r normalisation, echo=FALSE, warnings=FALSE, include=TRUE}

# Temporarily remove the lowly expressed genes
sce_filt <- sce[!(rowMeans(counts(sce)) <= 1 | fData(sce)$pct_dropout > 90),]

# Compute size factors without the lowly expressed genes
sf = computeSumFactors(counts(sce_filt), sizes=c(10,20,30,40), positive=TRUE)

# qplot(sf, sce_filt$total_counts, log="xy", ylab="Library size (mapped reads)", xlab="Size factor")
ggplot(data.frame(sf=log(sf), counts=log(sce_filt$total_counts))) +
  geom_point(aes(x=sf,y=counts)) +
  labs(y="Library size (log)", x="Size factor (log)") +
  theme_bw() +
  theme(
    axis.title = element_text(colour="black", size=15),
    axis.text = element_text(colour="black", size=12)
  )

# Normalise and log transform with the lowly expressed genes
sizeFactors(sce) <- sf; sce$sizeFactor <- sf
sizeFactors(sce_filt) <- sf; sce_filt$sizeFactor <- sf
sce <- normalize(sce, exprs_values="counts")
sce_filt <- normalize(sce_filt, exprs_values="counts")

# Update quality metrics
sce = calculateQCMetrics(sce)
```

<!-- # Mean vs variance plot -->
<!-- ```{r echo=TRUE, include=TRUE} -->
<!-- foo <- data.frame(sd=apply(exprs(sce),1,sd), mean=apply(exprs(sce),1,mean), spike=factor(isSpike(sce))) -->
<!-- ggplot(foo, aes(x=mean, y=sd)) + -->
<!--   geom_point(aes(color=spike)) + -->
<!--   stat_smooth() + -->
<!--   scale_color_manual(values=c("black","red")) + -->
<!--   xlab('Mean') + ylab('Standard deviation') -->
<!-- ``` -->


# Identify highly variable genes
We identify HVGs to focus on the genes that are driving heterogeneity across the population of cells. 
This requires estimation of the variance in expression for each gene, followed by decomposition of the variance into biological and technical components. HVGs are then identified as those genes with the largest biological components. 
This avoids prioritizing genes that are highly variable due to technical factors such as sampling noise during RNA capture and library preparation.
As suggest in (Lun et al), HVGs are defined as genes with biological components that are significantly greater than zero at a false discovery rate (FDR) of 5%. In addition, we only consider a gene to be a HVG if it has a biological component greater than or equal to 0.5. For transformed expression values on the log2 scale, this means that the average difference in true expression between any two cells will be at least 2-fold
```{r echo=FALSE, include=TRUE}

# Fit a mean-dependent trend to the gene-specific variances
var.fit <- trendVar(sce_filt, trend="loess", use.spikes=FALSE)

# Decompose the gene-specific variance into biological and technical components
# The technical component of the variance for each gene is determined by interpolating the fitted trend in fit at the mean log-CPM for that gene, which represents variance due to sequencing noise, variability in capture efficiency, etc. 
# The biological component is determined by subtracting the technical component from the total variance.
var.out <- decomposeVar(sce, var.fit)
var.out[is.nan(var.out$p.value),] <- NA
fData(sce)$techVar <- var.out$tech
fData(sce)$bioVar <- var.out$bio
fData(sce)$totalVar <- var.out$total

# Detect highly variable genes
# hvg.out <- var.out[which(var.out$bio >= 2.5),]
# hvg.out <- var.out[which(var.out$FDR <= 0.01),]
hvg.out <- var.out[which(var.out$FDR <= 0.01 & var.out$bio >= 1.0),]
hvg.out <- hvg.out[order(hvg.out$bio, decreasing=TRUE),]
fData(sce)$HVG <- rownames(sce) %in% rownames(hvg.out)
fData(sce)$HVG_high <- rownames(sce) %in% rownames(hvg.out) & fData(sce)$mean_exprs>5
fData(sce)$HVG_low <- rownames(sce) %in% rownames(hvg.out) & fData(sce)$mean_exprs<5

# Plot highly variable genes
plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", ylab="Variance of log-expression")
o <- order(var.out$mean)
lines(var.out$mean[o], var.out$tech[o], col="dodgerblue", lwd=2)
# points(var.out$mean[isSpike(sce_filt)], var.out$total[isSpike(sce_filt)], col="red", pch=16)
points(hvg.out$mean, hvg.out$total, col="green", pch=16)

# Detect lowly variable genes
lvg.out <- var.out[which(var.out$bio >= -0.5 & var.out$bio <= 0.5),]
lvg.out <- lvg.out[order(lvg.out$bio, decreasing=TRUE),]
fData(sce)$LVG <- rownames(sce) %in% rownames(lvg.out)
fData(sce)$LVG_high <- rownames(sce) %in% rownames(lvg.out) & fData(sce)$mean_exprs>5
fData(sce)$LVG_low <- rownames(sce) %in% rownames(lvg.out) & fData(sce)$mean_exprs<5

# Plot lowly variable genes
plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", ylab="Variance of log-expression")
o <- order(var.out$mean)
lines(var.out$mean[o], var.out$tech[o], col="dodgerblue", lwd=2)
# points(var.out$mean[isSpike(sce_filt)], var.out$total[isSpike(sce_filt)], col="red", pch=16)
points(lvg.out$mean, lvg.out$total, col="green", pch=16)

# print(sum(fData(sce)$HVG_low))
# print(sum(fData(sce)$HVG_high))
# print(sum(fData(sce)$LVG_low))
# print(sum(fData(sce)$LVG_high))

####
# pdf(file="/Users/ricard/NMT-seq/preprocessing/rna/out/meanvar.pdf", width=10, height=7, useDingbats=FALSE)
# plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", ylab="Variance of log-expression")
# o <- order(var.out$mean)
# lines(var.out$mean[o], var.out$tech[o], col="dodgerblue", lwd=2)
# dev.off()
###
```

We recommend checking the distribution of expression values for the top HVGs to ensure that the variance estimate is not being dominated by one or two outlier cells
```{r, echo=FALSE, include=TRUE}
plotExpression(sce, rownames(sce[fData(sce)$HVG_high == TRUE, ])[1:15]) 
plotExpression(sce, rownames(sce[fData(sce)$HVG_low == TRUE, ])[1:15]) 
plotExpression(sce, rownames(sce[fData(sce)$LVG_high == TRUE, ])[1:15]) 
plotExpression(sce, rownames(sce[fData(sce)$LVG_low == TRUE, ])[1:15]) 
```

```{r save, echo=FALSE, include=FALSE}
saveRDS(sce,io$out.file)
```
