---
title: "NMT-seq: general statistics of samples"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
```

```{r define_opts, echo=FALSE}

## Define I/O 
io <- list()
io$in.metadata <- "/Users/ricard/data/NMT-seq/sample_info.txt"
io$in.acc_data <- "/Users/ricard/data/NMT-seq/acc/raw/filtered/unstranded/binarised"
io$in.met_data <- "/Users/ricard/data/NMT-seq/met/raw/filtered/unstranded/binarised"
io$stats <- "/Users/ricard/NMT-seq/stats/samples/out/sample_stats.txt"

# io$basedir <- "/hps/nobackup/stegle/users/ricard/NMT-seq"
# io$in.metadata <- paste0(io$basedir,"/sample_info.txt")
# io$in.acc_data <- paste0(io$basedir,"/acc/raw/allele_unspecific/")
# io$in.met_data <- paste0(io$basedir,"/acc/raw/allele_unspecific/")
# io$stats <- "/homes/ricard/NOME-seq/stats/samples/sample_stats.txt"

## Define options
opts <- list()

# Define which cells to use
opts$cells <- fread(io$in.metadata)[,sample]
```

<!-- Load metadata -->
```{r}
metadata <- fread(io$in.metadata) %>% .[sample%in%opts$cells]
```

Run this only once to generate sample_stats.txt, then you can just load the file
```{r load_data_and_calculate_stats, echo=FALSE}
stats <- rbind(data.table(sample=opts$cells, context="CG", coverage=0, rate=0), data.table(sample=opts$cells, context="GC", coverage=0, rate=0))
for (cell in opts$cells) {
  # Met
  if (file.exists(sprintf("%s/%s.tsv.gz",io$in.met_data,cell))) {
    print(sprintf("Loading %s methylation...",cell))
    tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$in.met_data,cell), sep="\t", verbose=F, showProgress=F) %>% setnames(c("chr","pos","rate"))
    stats[sample==cell & context=="CG",coverage:=nrow(tmp)]
    stats[sample==cell & context=="CG",mean:=mean(tmp$rate)]
  } else {
    print(sprintf("Sample %s not found for methylation",cell))
  }

  # Acc
  if (file.exists(sprintf("%s/%s.tsv.gz",io$in.acc_data,cell))) {
    print(sprintf("Loading %s accessibility...",cell))
    tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$in.acc_data,cell), sep="\t", verbose=F, showProgress=F) %>% setnames(c("chr","pos","rate"))
    stats[sample==cell & context=="GC",coverage:=nrow(tmp)]
    stats[sample==cell & context=="GC",mean:=mean(tmp$rate)]
  } else {
    print(sprintf("Sample %s not found for accessibility",cell))
  }
}

fwrite(stats,file=io$stats, sep="\t", row.names=F, col.names=T)
```


```{r read_stats, echo=FALSE}
stats <- fread(io$stats) %>% .[sample %in% opts$cells] %>%
  merge(metadata,by="sample")
```

<!-- Number of observed CG/GC sites per sample (ignore, this is plotted in QC_accmet.Rmd) -->
<!-- ```{r coverage1, echo=FALSE} -->
<!-- opts$met_coverage_threshold <- 2e5 -->
<!-- opts$acc_coverage_threshold <- 1e6 -->
<!-- foo <- stats[,c("sample","coverage","context")] -->
<!-- cellcolor <- c("black","red")[as.numeric(foo$coverage<opts$met_coverage_threshold | foo$coverage<opts$acc_coverage_threshold )+1] -->
<!-- p <- ggplot(foo, aes(x=sample, y=coverage, fill=context)) + -->
<!--   geom_bar(stat='identity', position="dodge") + -->
<!--   scale_y_continuous(expand=c(0,0)) + -->
<!--   scale_fill_discrete(labels=c("CG methylation","GC accessibility")) + -->
<!--   geom_hline(yintercept=opts$met_coverage_threshold, colour="#F8766D", linetype="dashed") + -->
<!--   geom_hline(yintercept=opts$acc_coverage_threshold, colour="#00BFC4", linetype="dashed") + -->
<!--   ylab("Number of observed sites") + -->
<!--   theme( -->
<!--     axis.title.y = element_text(colour="black", size=17, margin=margin(0,20,0,0)), -->
<!--     axis.title.x = element_blank(), -->
<!--     axis.text.x = element_text(colour="black", color=cellcolor, angle=90, size=10, vjust=0.5, hjust=1.0), -->
<!--     axis.text.y = element_text(colour="black", size=11), -->
<!--     axis.ticks = element_line(colour="black"), -->
<!--     axis.line = element_blank(), -->
<!--     legend.position="top", -->
<!--     legend.title = element_blank(), -->
<!--     legend.direction = "horizontal", -->
<!--     legend.key.width=unit(1.2,"line"), -->
<!--     legend.key.height=unit(1.0,"line"), -->
<!--     legend.text = element_text(size=15), -->
<!--     panel.background = element_blank(), -->
<!--     panel.grid = element_blank(), -->
<!--     panel.border = element_blank() -->
<!--     ) -->
<!-- p -->
<!-- pdf("/Users/ricard/NMT-seq/rebuttal/EB/stats/out/coverage.pdf", height=5, width=7) -->
<!-- print(p) -->
<!-- dev.off() -->
<!-- ``` -->

<!-- Mean methylation/accessibility rate per sample -->
```{r mean, echo=FALSE}

# Methylation
foo <- stats[context=="CG" & pass_metQC==T,c("sample","mean")]
foo %>% setkey(mean) %>% .[,sample:=factor(sample,levels=sample)]

p <- ggplot(foo, aes(x=sample, y=mean)) +
  geom_bar(stat='identity', position="dodge", fill="red") +
  scale_y_continuous(expand=c(0,0)) +
  ylab("Genome-wide mean methylation rate") +
  theme(
    axis.title.y = element_text(colour="black", size=17, margin=margin(0,20,0,0)),
    axis.title.x = element_blank(),
    axis.text.x = element_text(colour="black", angle=90, size=10, vjust=0.5, hjust=1.0),
    axis.text.y = element_text(colour="black",size=rel(1.5)),
    axis.ticks = element_line(colour="black"),
    axis.line = element_line(color="black"),
    legend.position="top",
    legend.title = element_blank(),
    legend.direction = "horizontal",
    legend.key.width=unit(1.2,"line"),
    legend.key.height=unit(1.0,"line"),
    legend.text = element_text(size=15),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    panel.border = element_blank()
  )
p

# Accessibility
foo <- stats[context=="GC" & method=="NMT" & pass_accQC==T,c("sample","mean")]
foo %>% setkey(mean) %>% .[,sample:=factor(sample,levels=sample)]
p <- ggplot(foo, aes(x=sample, y=mean)) +
  geom_bar(stat='identity', position="dodge", fill="blue") +
  scale_y_continuous(expand=c(0,0)) +
  ylab("Genome-wide mean accessibility rate") +
  theme(
    axis.title.y = element_text(colour="black", size=17, margin=margin(0,20,0,0)),
    axis.title.x = element_blank(),
    axis.text.x = element_text(colour="black", angle=90, size=10, vjust=0.5, hjust=1.0),
    axis.text.y = element_text(colour="black",size=rel(1.5)),
    axis.ticks = element_line(colour="black"),
    axis.line = element_line(color="black"),
    legend.position="top",
    legend.title = element_blank(),
    legend.direction = "horizontal",
    legend.key.width=unit(1.2,"line"),
    legend.key.height=unit(1.0,"line"),
    legend.text = element_text(size=15),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    panel.border = element_blank()
  )
p

# foo <- stats[,c("sample","mean","context")]
# p <- ggplot(foo, aes(x=sample, y=mean, fill=context)) +
#   geom_bar(stat='identity', position="dodge") +
#   scale_y_continuous(expand=c(0,0)) +
#   scale_fill_discrete(labels=c("CG methylation","GC accessibility")) +
#   ylab("Genome-wide mean rate") +
#   theme(
#     axis.title.y = element_text(colour="black", size=17, margin=margin(0,20,0,0)),
#     axis.title.x = element_blank(),
#     axis.text.x = element_text(colour="black", angle=90, size=9, vjust=0.5, hjust=1.0),
#     axis.text.y = element_text(colour="black",size=rel(1.5)),
#     axis.ticks = element_line(colour="black"),
#     axis.line = element_line(color="black"),
#     legend.position="top",
#     legend.title = element_blank(),
#     legend.direction = "horizontal",
#     legend.key.width=unit(1.2,"line"),
#     legend.key.height=unit(1.0,"line"),
#     legend.text = element_text(size=15),
#     panel.background = element_blank(),
#     panel.grid = element_blank(),
#     panel.border = element_blank()
#   )
# p

# pdf("/Users/ricard/NMT-seq/stats/samples/out/means.pdf", height=5, width=7)
# print(p)
# dev.off()
```
