---
title: "NMT-Seq: Plot lineage determination from accessibility profiles and methylation"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(BPRMeth))
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(data.table))
```

```{r echo=FALSE, include=FALSE}
ggplot_bpr_profiles <- function(X, obs, title="", subtitle="", up_label="-500bp", down_label="+500bp", middle_label="TSS", ...){
  # ggplot(X, aes(x = xs, y = ys)) +
  p <- ggplot(data = data.frame(X), aes(x = xs, y = ys)) +
    geom_line(aes(x = xs, y = ys), size=1.5, col = "darkblue") +
    geom_point(data = obs, mapping = aes(x = x, y = y), shape=1, color="red", size=3) +
    geom_ribbon(data = data.frame(X), mapping = aes(ymin= ys_low, ymax= ys_high), alpha=0.25, size = 0.1, fill="cornflowerblue") +
    scale_x_continuous(limits = c(-1, 1), labels =c(up_label, "", middle_label, "", down_label)) + 
    scale_y_continuous(limits = c(0, 1), breaks=pretty_breaks(n=4)) + 
    labs(title = title, subtitle = subtitle, x="genomic region", y="Accessibility") + 
    line_theme()# + theme(axis.text.x = element_blank())
}

ggplot_bpr_profiles_metacc <- function(X_acc, obs_acc, X_met, obs_met, title="", subtitle="", up_label="-500bp", down_label="+500bp", middle_label="TSS", ...){
  # ggplot(X, aes(x = xs, y = ys)) +
  p <- ggplot(data = data.frame(X_acc), aes(x = xs, y = ys)) +
    geom_line(aes(x = xs, y = ys), size=1.5, col = "darkblue") +
    geom_point(data = obs_acc, mapping = aes(x = x, y = y), shape=1, color="blue", size=3) +
    geom_ribbon(data = data.frame(X_acc), mapping = aes(ymin= ys_low, ymax= ys_high), alpha=0.4, size = 0.1, fill="cornflowerblue") +
    geom_line(data = data.frame(X_met), mapping = aes(x = xs, y = ys), size=1.5, col = "red3") +
    geom_point(data = obs_met, mapping = aes(x = x, y = y), shape=1, color="red", size=3) +
    geom_ribbon(data = data.frame(X_met), mapping = aes(ymin= ys_low, ymax= ys_high), alpha=0.4, size = 0.1, fill="coral") +
    scale_x_continuous(limits = c(-1, 1), labels =c(up_label, "", middle_label, "", down_label)) + 
    scale_y_continuous(limits = c(0, 1), breaks=pretty_breaks(n=4)) + 
    labs(title = title, subtitle = subtitle, x="genomic region", y="GC / CG level") + 
    line_theme()# + theme(axis.text.x = element_blank())
}

# Function for plotting BPR profiles
plot_ml_bpr_profiles <- function(region, X, fit_prof, thresh = 0.9,
                              title = "Gene promoter", up_label = "-500bp",
                              down_label = "+500bp", middle_label = "TSS", ...){
  #graphics::par(cex=1.05, mai=c(0.87, 0.85, .7, .3) )
  xs <- seq(from = -1, to = 1, by = 0.01)
  graphics::plot(X[[region]][, 1], X[[region]][, 2], col = "blue2", pch = 21, ylim = c(0,1),
                 xlim = c(-1,1), lwd = 0.8, xlab = NA, ylab = NA,
                 cex.axis = 1.4, xaxt = "n")
  #graphics::mtext(side = 1, "genomic region", line = 3, cex = 1.2)
  graphics::mtext(side = 2, "GC accessibility", line = 3, cex = 1.2)
  graphics::axis(side = 1, at = c(-1, 0, 1), labels=c(up_label, middle_label, down_label), cex.axis = 1.4)
  graphics::title(main=title, line = 1, cex.main=1.8)
  graphics::lines(x = xs, y = eval_probit_function(fit_prof$basis, xs, fit_prof$W_opt[region, 1:(fit_prof$basis$M + 1)]), col = "red2", lwd = 2)
}

# Define ggplot2 theme for scatter plots
scatter_theme <- function(){
  p <- theme(
      plot.title=element_text(size=28, face='bold', margin=margin(0,0,10,0), hjust=0.5),
      plot.subtitle = element_text(size=20, face='bold', margin=margin(0,0,10,0), hjust=0.5),
      axis.text=element_text(size=rel(1.75), color='black'),
      axis.title=element_text(size=rel(1.95), color='black'),
      axis.title.y = element_text(margin=margin(0,10,0,0)),
      axis.title.x = element_text(margin=margin(10,0,0,0)),
      legend.position="right",
      panel.border=element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank()
    )
}

# Scatter plot function
scatter_plot <- function(data, xlabel="", ylabel="", title="", subtitle="", lm=FALSE, loess = FALSE) {
  p <- ggplot(data, aes_string(x=names(data)[1], y=names(data)[2], col=names(data)[3])) +
    labs(x=xlabel, y=ylabel, title=title, subtitle=subtitle) +
    geom_point(shape = 16, size = 2.5, show.legend = FALSE, alpha = .55) + 
    scatter_theme() + 
    scale_color_gradient(low = "#0091ff", high = "#f0650e")
  if (lm){
    if (loess){ p <- p + geom_smooth() }#+ scale_x_continuous(trans = "log1p")
    else{ p <- p + geom_smooth(method='lm') }
  }
}

# Define ggplot2 boxplot theme for boxplots
boxplot_theme <- function() {
  p <- theme(
    plot.title = element_text(size=30, hjust=0.5, margin=margin(0,0,20,0)),
    axis.title.y = element_text(colour="black", size=20, vjust=1.5),
    axis.title.x = element_text(colour="black", size=20, vjust=1.5, margin=margin(15,0,0,0)),
    axis.text.x = element_text(colour="black",size=rel(1.6)),
    axis.text.y = element_text(colour="black",size=rel(1.6)),
    axis.line = element_line(colour="black", size=rel(0.7)),
    axis.ticks.x = element_line(colour="black", size=rel(0.8)),
    axis.ticks.y = element_blank(),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    legend.position="none",
    legend.text=element_text(size=15),
    legend.title=element_blank(),
    legend.background=element_blank(),
    panel.border = element_blank()
  )
}

# Define ggplot2 theme for line plots
line_theme <- function(){
  p <- theme(
      plot.title=element_text(size=25, face='bold', margin=margin(0,0,5,0), hjust=0.5),
      axis.text=element_text(size=rel(1.15), color='black'),
      axis.title=element_text(size=rel(1.55), color='black'),
      axis.title.y = element_text(margin=margin(0,15,0,0)),
      axis.title.x = element_text(margin=margin(15,0,0,0)),
      axis.ticks.x = element_line(colour="black", size=rel(0.8)),
      axis.ticks.y = element_blank(),
      legend.position="right",
      legend.key.size = unit(1.9, 'lines'),
      legend.title=element_text(size=24, face='bold'),
      legend.text=element_text(size=19),
      panel.border=element_blank(),
      panel.grid.major = element_line(colour = "gainsboro"),
      #panel.grid.minor = element_line(colour = "grey"),
      panel.background = element_blank()
    )
}

#' Align cluster indices
#' 
#' \code{align_clusters} function helps to align cluster indices after each MCMC
#' simulation for each data source.
#' 
#' @param Z1 Previous cluster assignments.
#' @param Z2 Current cluster assignments
#' @param type Object type of the cluster assignemnts, either 'vector' or
#'   'matrix'.
#'   
#' @return The aligned indices of the current cluster assignments.
#' 
align_clusters <- function(Z1, Z2, type = "vec"){
  if(type == "vec"){
    for(k in 1:length(unique(Z1))){ # For each cluster k in previous Cluster
      # Find Max 
      Max <- sum(Z1==k & Z2==k)/(.01 + sum(Z1==k) + sum(Z2==k))
      for(tempk in  1:length(unique(Z2))){ # For each cluster k in current Cluster
        # Check if the proportions are higher than Max
        if( (sum(Z1==k & Z2==tempk)/(.01 + sum(Z1==k) + sum(Z2==tempk))) > Max){
          # Get the proportion that the two cluster indices are the same
          Max <- sum(Z1==k & Z2==tempk)/(.01 + sum(Z1==k) + sum(Z2==tempk))
          dummy <- (Z2==k)      # Keep indices that do not match
          Z2[Z2==tempk] <- k    # Swap the incorrect indices
          Z2[dummy] <- tempk    # Swap the incorrect indices
        }
      }
    }
  }else if(type == "mat"){
    for(k in 1:dim(Z1)[2]){         # For each cluster k in previous Cluster
      for(tempk in  1:dim(Z2)[2]){  # For each cluster k in current Cluster
        Max <- sum(Z1==Z2)          # Number of matches between the cluster indices
        Z2dummy <- Z2               # Keep the current indices in a dummy variable
        Z2dummy[,k] = Z2[,tempk]    # Swap the incorrect indices
        Z2dummy[,tempk] = Z2[,k]    # Swap the incorrect indices
        if(sum(Z1==Z2dummy) > Max){ # If the swaps make a better alignment, update indices
          Z2 <- Z2dummy
        }
      }
    }
  }
  return(Z2) # Return the aligned cluster indices
}
```

<!-- # Parse and filter data -->
```{r echo=FALSE, include=FALSE}
# Data
io                     <- list()
io$base_dir            <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data-eb"
# Load methylation data
load(paste0(io$base_dir,"/met/parsed/profiles/met_400bp.RData"))
region_dt_met <- region_dt
# Load accessibility data
load(paste0(io$base_dir,"/acc/parsed/profiles/acc_400bp.RData"))
region_dt_acc <- region_dt
io$base_dir            <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data-eb"
io$out_dir             <- "/home/cakapourani/Documents/Code/nmt-seq/rebuttal/EB/lineage_profiles/out/"
io$rna_file            <- paste0(io$base_dir, "/rna/parsed/sceset.rds")
io$acc_file            <- paste0(io$base_dir, "/acc/parsed/profiles/lineage_cluster_prom_200_200_basis13_GpCcov8_cellcov0.7.rds")
io$meta_file           <- paste0(io$base_dir, "/sample_sheet.csv")
io$genelist_dir        <- paste0(io$base_dir, "/metadata")
io$plur_genes_file     <- paste0(io$genelist_dir, "/pluripotency_extended.tsv")
io$top_hits            <- paste0(io$out_dir, "/agreement_list.rds")
opts$top_N             <- 80
opts$basis_prof_met    <- create_rbf_object(M = 9)
rm(region_dt)
```

<!-- # Load scRNA data -->
```{r load_metadata, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
# Load cell metadata
metadata <- fread(io$meta_file) %>% .[,c("sample", "culture", "pass_metQC", "pass_accQC", "pass_rnaQC", "lineage")] %>% 
  na.omit() %>% .[pass_metQC == TRUE & pass_accQC ==  TRUE & pass_rnaQC == TRUE]
```

<!-- # Load scRNA data -->
```{r load_rna, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
sceset       <- readRDS(file = io$rna_file)  # Load expression as SCESet object
# Subset cells
sceset <- sceset[, metadata$sample]
# Create data.table
rna_dt       <- exprs(sceset) %>% t %>% as.data.table(keep.rownames = "sample") %>% melt(id.vars = "sample", value.name = "expr", variable.name = "gene")

# Extract gene coordinates metadata to do the overlap
rna_metadata <- fData(sceset) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% .[,c("chr", "start", "end", "gene", "ens_id")]  %>% .[, chr := as.factor(sub("chr", "", chr))] %>% setnames("ens_id", "id")

rna_dt       <- merge(rna_metadata[, c("chr", "id", "gene")], rna_dt, by = c("gene"))     # Merge using only samples that pass filtering
rm(sceset, rna_metadata)
```

<!-- # Load sc-NOMe-seq data -->
```{r load_acc, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
accrna_dt  <- readRDS(io$top_hits)
```


# Specific gene examples
```{r ggg, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}

for (hits in 1:opts$top_N){
  gene_id <- accrna_dt$id[hits]                       # Get gene ENS id
  gene_dt_acc <- lapply(region_dt_acc, "[[", gene_id) # Extract accessibility data
  gene_dt_met <- lapply(region_dt_met, "[[", gene_id) # Extract accessibility data
  cl_names <- accrna_dt$cell_names[[hits]]            # Names of cells passed filtering
  cl_assign <- accrna_dt$cell_clusters[[hits]]        # Cluster assignment of cells
  cl_profiles <- accrna_dt$cell_profiles[[hits]]      # Learned profiles for each cluster
  K <- NCOL(cl_profiles)                              # Number of clusters K
  xs <- seq(-1, 1, length = 100)
  ys <- matrix(0, ncol = K, nrow = length(xs))
  labs <- vector("character", length = K)
  opts$basis_prof <- create_rbf_object(M = NROW(cl_profiles) - 1)
  for (k in 1:K){   # Iterate over each cluster
    ys[,k] <- eval_probit_function(opts$basis_prof, xs, cl_profiles[, k]) # Evaluate function for plotting
    nnn <- cl_names[which(cl_assign == k)]                                # Get names of cells belonging to each cluster
    labs[k] <- paste0("C", k, " ", round(median(rna_dt[id == gene_id & sample %in% nnn]$expr), 2))
  }
  # Create a data.table object
  dt2 <- data.table(ys) %>% setnames(labs) %>% melt(variable.name = "Cluster", value.name = "ys") %>% .[, xs := xs]
  # Create plot of cluesters
  p_cl_prof <- ggplot(dt2, aes(x = xs, y = ys, color=Cluster)) + geom_line(size=2) +
      scale_x_continuous(limits = c(-1, 1), labels =c("-200bp", "", "TSS", "", "+200bp")) + 
      scale_y_continuous(limits = c(0, 1), breaks=pretty_breaks(n=4)) + 
      scale_color_brewer(palette="Dark2") + scale_fill_brewer(palette="Dark2") +
      labs(title = paste0("Gene ", accrna_dt$gene[hits]), x="genomic region", y="Accessibility") + line_theme()
  
  # Create subdirectories for each gene
  io$out_dir_loc <- paste0(io$out_dir, "/metacc/low_cov/", hits, "_", gene_id, "/")
  ifelse(!dir.exists(file.path(io$out_dir_loc)), dir.create(file.path(io$out_dir_loc)), FALSE)
  
  cluster_plots <- list()
  for (k in 1:K){
    # Get genes assigned to cluster k accessibility data
    gene_dt_acc_filt <- gene_dt_acc[cl_names[which(cl_assign == k)]]
    cells_names <- names(gene_dt_acc_filt)
    cl_pop <- length(cells_names)  # Number of cells in cluster k
    
    # Get gene profiles for methylation data
    gene_dt_met_filt <- gene_dt_met[cells_names]
    gene_dt_met_filt <- gene_dt_met_filt[which(!is.na(gene_dt_met_filt))]
    # Learn profiles for each gene accessibility
    prof_cells_acc <- bpr_bayes(x = gene_dt_acc_filt[1:cl_pop], basis = opts$basis_prof, w_0_cov = diag(4, opts$basis_prof$M + 1), gibbs_nsim = 3000, 
                            gibbs_burn_in = 1000, is_parallel = FALSE, keep_gibbs_draws = TRUE)
    if (length(gene_dt_met_filt) > 0){
      # Learn profiles for each gene methylation
      prof_cells_met <- bpr_bayes(x = gene_dt_met_filt, basis = opts$basis_prof_met, w_0_cov = diag(4, opts$basis_prof_met$M + 1), gibbs_nsim = 3000, 
                            gibbs_burn_in = 1000, is_parallel = FALSE, keep_gibbs_draws = TRUE)
    }else{
      prof_cells_met <- list()
    }
    
    
    H_acc <- design_matrix(x = opts$basis_prof, obs = xs)$H  # Create design matrix
    H_met <- design_matrix(x = opts$basis_prof_met, obs = xs)$H  # Create design matrix
    ys_draws <- matrix(0, nrow = NROW(prof_cells_acc$W_draws[[1]]), ncol = length(xs))
    
    # Initialize data
    dt_acc <- data.table(xs = xs, ys = 0, ys_low = 0, ys_high = 0)
    dt_met <- data.table(xs = xs, ys = 0, ys_low = 0, ys_high = 0)
    pp <- list()
    met_k <- 1
    for (m in 1:cl_pop){
      # Accessibility predictive distribution
      w_draws <- prof_cells_acc$W_draws[[m]] # Compute predictive distribution
      for (i in 1:NROW(prof_cells_acc$W_draws[[1]])){ ys_draws[i, ] <- pnorm(H_acc %*% w_draws[i, ]) }
      # Compute quantiles of ys
      ys_q <- apply(ys_draws, 2, quantile, probs = c(0.1, 0.9),  na.rm = TRUE)
      dt_acc <- dt_acc %>% .[, c("ys", "ys_low", "ys_high") := list(colMeans(ys_draws), ys_q[1, ], ys_q[2, ])]
      obs_acc <- as.data.table(gene_dt_acc_filt[[m]]) %>% setnames(c("x", "y"))
      
      if (!is.null(gene_dt_met_filt[[cells_names[m]]])){
        # Methylation predictive distribution
        w_draws <- prof_cells_met$W_draws[[met_k]] # Compute predictive distribution
        for (i in 1:NROW(prof_cells_met$W_draws[[1]])){ ys_draws[i, ] <- pnorm(H_met %*% w_draws[i, ]) }
        # Compute quantiles of ys
        ys_q <- apply(ys_draws, 2, quantile, probs = c(0.1, 0.9),  na.rm = TRUE)
        dt_met <- dt_met %>% .[, c("ys", "ys_low", "ys_high") := list(colMeans(ys_draws), ys_q[1, ], ys_q[2, ])]
        obs_met <- as.data.table(gene_dt_met_filt[[met_k]]) %>% setnames(c("x", "y"))
        met_k <- met_k + 1
        pp[[m]] <- ggplot_bpr_profiles_metacc(X = dt_acc, obs = obs_acc, X_met = dt_met, obs_met = obs_met, title = paste0("Cell ", cells_names[m]), subtitle = paste0("Expr = ", round(rna_dt[id == gene_id & sample == cells_names[m], expr], 3)),  up_label = "-200bp", down_label = "+200bp", middle_label = "TSS")
      }else{
        pp[[m]] <- ggplot_bpr_profiles(X = dt_acc, obs = obs_acc, title = paste0("Cell ", cells_names[m]), subtitle = paste0("Expr = ", round(rna_dt[id == gene_id & sample == cells_names[m], expr], 3)),  up_label = "-200bp", down_label = "+200bp", middle_label = "TSS")
      }
      
      pdf(file = paste0(io$out_dir_loc, "c", k, "_", metadata[sample == cells_names[m]]$lineage, "_", cells_names[m], "_", gene_id, ".pdf"), width = 6, height = 4)
      print(pp[[m]])
      dev.off()
    }
    # cluster_plots[[k]] <- cowplot::plot_grid(plotlist = pp, label_size = 30, ncol = 2, nrow = 4, rel_widths = c(1, 1.2))
    
    # # ML profiles
    # prof_cells <- bpr_optim(x = gene_dt_acc_filt[1:cl_pop], basis = opts$basis_prof, lambda = 1/20, fit_feature = NULL, cpg_dens_feat = FALSE, opt_itnmax = 50, is_parallel = FALSE)
    # 
    # for (m in 1:cl_pop){
    #   pdf(file = paste0(io$out_dir_loc, "c", k, "_", metadata[sample == cells_names[m]]$lineage, "_", cells_names[m], "_", gene_id, "_ML.pdf"), width = 6, height = 4)
    #   plot_ml_bpr_profiles(region = m, X = gene_dt_acc_filt, fit_prof = prof_cells, title = paste0("Cell ", cells_names[m]), 
    #                 up_label = "-200bp", down_label = "+200bp", middle_label = "TSS")
    #   dev.off()
    # }
  }
  
  pdf(file = paste0(io$out_dir_loc, "cl_profiles_", gene_id, ".pdf"), width = 6, height = 4)
  print(p_cl_prof)
  dev.off()
}
```

