---
title: "DNase_promoters"
author: "Stephen Clark"
date: "04/04/2017"
output: html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(data.table)
library(purrr)
library(Rsamtools)
library(ggplot2)
library(cowplot)

io = list(dhs_bam = "data/dhs/129_DHS_bamfile.bam",
          gene_anno = "data/feat/ricard/genebody.bed",
          raw_acc = "data/raw_dna/"
          )

opts = list(prom_size = 2000,
            win = 25
           )

```
# Encode DNase-seq data  

Read in BAM files, count number of reads at each position and subset to promoters.
Next, measure the distance of each position to the nearest TSS then find the mean number of reads for any given bp relative to the nearest TSS. Plot this data.  

```{r}

anno = fread(io$gene_anno, select = c(1:3, 6)) %>% 
  setnames(c("chr", "start", "end", "strand")) %>% 
  .[, chr := gsub("chr", "", chr)] %>% 
  setkey(strand) %>% 
  # promoter is defined as tss (start of gene on plus or end on minus) +/- opts$prom_size
  .["+", tss := start] %>% 
  .["-", tss := end] %>% 
  .[, start := tss - opts$prom_size] %>% 
  .[, end := tss + opts$prom_size]



dhs = scanBam(io$dhs_bam, param = ScanBamParam(what = c("rname", "pos", "qwidth"))) %>% 
  .[[1]] %>% 
  as.data.table() %>% 
  setnames(c("chr", "start", "end")) %>% 
  .[, chr := gsub("chr", "", chr)] %>% 
  .[, end := start + end] %>%  # change end column from width to end position
  .[, .N, .(chr, start, end)] %>%  # count number of reads at each position
  .[complete.cases(.)] %>% 
  setkey(chr, start, end) %>% 
  foverlaps(anno %>% setkey(chr, start, end), nomatch = 0) %>% 
  setkey(strand) %>% 
  .["+", bp := i.start - tss] %>% 
  .["-", bp := tss - i.start] %>% 
  .[, mean(N), bp] %>% 
  .[, dataset := "DNase-seq"]



ggplot(dhs, aes(bp, V1)) + geom_line() + xlim(-1000, 1000)
```

# scNMT-seq data  

Load all GpC methylation files for cells which pass QC. 

```{r}

acc = list.files(io$raw_acc, pattern = ".tsv", full.names = TRUE) %>% 
  map(~fread(.) %>%
        setnames(c("chr", "start", "rate")) %>% 
        .[, end := start] %>% 
        setkey(chr, start, end) %>% 
        foverlaps(anno %>% setkey(chr, start, end), nomatch = 0) %>% 
        # first average rate by position in each cell individually
        .[, .(rate = mean(rate)), .(chr, start, end, strand, tss, i.start)]
  ) %>% 
  rbindlist() %>% 
  # now average each position accross cells
  .[, .(rate = mean(rate)), .(chr, start, end, strand, tss, i.start)] %>% 
  setkey(strand) %>% 
  .["+", bp := i.start - tss] %>% 
  .["-", bp := tss - i.start] %>% 
  # and find average rate relative to TSS
  .[, mean(rate), bp] %>% 
  .[, dataset := "NMT-seq"]



ggplot(acc, aes(bp, V1)) + geom_line() + xlim(-1000, 1000)

```


Now combine datasets and re-calculate rates as percentage of the maximum in that dataset, so as to enable plotting count based data with % methylation data. Additionally rates are average over windows (`r opts$win`bp) for smoothing.  

```{r}
plots = rbind(acc, dhs) %>% 
  .[, V1 := V1/max(V1), dataset] %>% 
  .[, bp := opts$win * round(bp /opts$win)] %>% 
  .[, mean(V1), .(dataset, bp)]

p = ggplot(plots, aes(bp, V1)) +
  stat_summary(aes(group = dataset, colour = dataset, fill = dataset, linetype = dataset),  geom="smooth") + 
    ggtitle("Promoters") +
    xlim(-opts$prom_size, opts$prom_size) +
    xlab("Genomic distance from TSS") +
    ylab("Scaled accessibility rate") +   
    scale_linetype_manual(values=c("NMT-seq"="solid", "DNase-seq"="solid")) +
    scale_colour_manual(values=c("DNase-seq"="#00BFC4", "NMT-seq"="#F8766D"), labels=c("Encode DNase-seq", "Merged scNMT-seq")) +
    guides(fill=FALSE, linetype=FALSE) +
    theme(
      plot.margin = unit(c(t=1,r=1,b=1,l=1), "cm"),
      plot.title = element_text(size=25,hjust=0.5),
      axis.text=element_text(size=16, colour="black"),
      axis.title.x=element_text(size=17, margin=margin(10,0,0,0)),
      axis.title.y=element_text(size=17, margin=margin(0,10,0,0)),
      axis.line = element_line(size=rel(1.0)),
      axis.ticks = element_line(size=rel(1.5), color="black"),
      legend.key = element_blank(),
      legend.position = "top",
      legend.direction = "horizontal",
      # legend.key.size= unit(0.5, "cm"),
      legend.key.width=unit(1.2,"line"),
      legend.key.height=unit(1.0,"line"),
      legend.margin = margin(t=10, r=0, b=0, l=0, unit="pt"),
      legend.title = element_blank(),
      legend.text = element_text(size=17),
      panel.border=element_blank(),
      panel.grid.major=element_blank(),
      panel.grid.minor=element_blank(),
      panel.background=element_blank()
    )
p
save_plot("plots/dnase_profile.pdf", p, base_width = 8, base_height = 8)
```