---
title: "NMT-seq EB: profiles of DNA methylation and accessibility"
output: 
  BiocStyle::html_document:
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
library(scater)
library(data.table)
library(purrr)
library(ggplot2)
```

```{r define_opts, echo=FALSE, include=FALSE}

io <- list()
io$basedir <- "/Users/ricard/data/NMT-seq_EB"
io$outdir <- "/Users/ricard/NMT-seq/rebuttal/EB/profiles/out"
io$features.indir <- paste(io$basedir,"features/filt",sep="/")
io$in.sample_metadata <- paste0(io$basedir,"/sample_sheet.csv")
io$met.indir <- paste(io$basedir,"met/raw/merged/binarised",sep="/")
io$acc.indir <- paste(io$basedir,"acc/raw/merged/binarised",sep="/")
io$rna.infile <- paste(io$basedir,"rna/parsed/sceset.rds",sep="/")

opts <- list()
opts$window_size <- 1000  # symmetric window size to display (up and downstream)
opts$tile <- 50           # window size to calculate rates
# opts$gene_window <- 2000  # for the overlap between genes and non-genic genomic contexts

# Define which cells to use
opts$cells <- fread(io$in.sample_metadata, header=T) %>% .[pass_metQC==T & pass_accQC==T,sample]
# opts$cells <- c("A07","A08","A09")

# Define which annotations to use and where to center the window (start, end or center)
opts$annotations <- c(
  "prom_500_500"="center",
  # "prom_500_500_cgi"="center",
  # "prom_500_500_noncgi"="center",
  # "prom_500_500_active"="center",
  # "prom_500_500_inactive"="center",
  # "prom_500_500_poised"="center",
  "active_enhancers"="center",
  "super_enhancers"="center",
  "primed_enhancers"="center",
  "CTCF"="center",
  "p300"="center",
  "Nanog"="center",
  "Oct4"="center"
  )
```

<!-- ```{r print_opts, echo=FALSE, include=TRUE} -->
<!-- cat(sprintf("- Cells: %s\n",paste(opts$cells,collapse=" "))) -->
<!-- cat(sprintf("- Window size: %d\n",opts$window_size)) -->
<!-- cat(sprintf("- Tile: %d\n",opts$tile)) -->
<!-- cat(sprintf("- Gene extension upstream and downstream for the overlap: %d\n",opts$gene_window)) -->
<!-- ``` -->

<!-- Load genomic contexts and define windows -->
```{r load_features, echo=FALSE, include=FALSE}
anno_list <- list()
for (anno in names(opts$annotations)) {
  tmp <- fread(sprintf("%s/%s.bed",io$features.indir,anno))[,c(1,2,3,4,5,6)]
  colnames(tmp) <- c("chr","start","end","strand","id","anno")
  
  # Define central position for the window approach
  if (opts$annotations[anno] == "start") {
    tmp <- rbind(tmp[strand=="+",.(chr,start,strand,id,anno)] %>% .[,center:=start] %>% .[,c("start"):=NULL], 
                 tmp[strand=="-",.(chr,end,strand,id,anno)] %>% .[,center:=end] %>% .[,c("end"):=NULL]) 
  }
  if (opts$annotations[anno] == "center") {
    stopifnot(all(tmp[,end] > tmp[,start]))
    tmp <- tmp[,.(chr,start,end,strand,id,anno)][,center:=round(end+start)/2][,c("start","end"):=NULL]
  }
  if (opts$annotations[anno] == "end") {
    tmp <- rbind(tmp[strand=="+",.(chr,end,strand,id,anno)][,center:=end][,c("end"):=NULL], 
                 tmp[strand=="-",.(chr,start,strand,id,anno)][,center:=start][,c("start"):=NULL])
  }
  anno_list[[anno]] <- tmp %>% .[, c("start","end") := list(center-opts$window_size,center+opts$window_size)]
}

anno_df <- rbindlist(anno_list) %>% 
  .[,c("anno","chr"):=list(as.factor(anno),as.factor(sub("chr","",chr)))] %>%
  setkey(chr,start,end)
rm(anno_list)
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$in.sample_metadata) %>% .[sample%in%opts$cells]
```

<!-- Load methylation and accessibility data -->
```{r load_accmet, echo=FALSE, include=FALSE}

# Read accessibility data
acc_list <- list()
for (cell in opts$cells) {
  tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$acc.indir,cell), sep="\t", showProgress=F) %>%
    setnames(c("chr","pos","rate")) %>% .[,sample:=cell] %>% .[,c("start","end"):=list(pos,pos)] %>% setnames("pos","bp") %>% 
    setkey("chr","start","end") %>%
    foverlaps(.,anno_df, nomatch=0) %>% .[, c("chr","i.start","i.end") := NULL] %>%
    .[,dist:=ifelse(strand %in% c("+","*"),bp-center,center-bp)] %>% 
    .[, dist:=opts$tile*round(dist/opts$tile)] %>%
    .[,list(rate=mean(rate), n=.N),by=.(sample,id,dist,anno)]
  acc_list[[cell]] <- tmp
}
acc <- rbindlist(acc_list) %>% .[,context:="GC"]
rm(acc_list)

# Read methylation data
met_list <- list()
for (cell in opts$cells) {
  tmp <- fread(sprintf("zcat < %s/%s.tsv.gz",io$met.indir,cell), sep="\t", showProgress=F) %>%
    setnames(c("chr","pos","rate")) %>% .[,sample:=cell] %>% .[,c("start","end"):=list(pos,pos)] %>% setnames("pos","bp") %>% setkey("chr","start","end") %>%
    foverlaps(.,anno_df, nomatch=0) %>% .[, c("chr","i.start","i.end") := NULL] %>%
    .[,dist:=ifelse(strand %in% c("+","*"),bp-center,center-bp)] %>% 
    .[, dist:=opts$tile*round(dist/opts$tile)] %>%
    .[,list(rate=mean(rate), n=.N),by=.(sample,id,dist,anno)]
  met_list[[cell]] <- tmp
}
met <- rbindlist(met_list) %>% .[,context:="CG"]
rm(met_list)

# Combine met and acc and remove old data to free some space
metacc <- rbind(met,acc)
rm(acc,met)

```

<!-- Perform basic filtering on the data -->
<!-- ```{r filtering, echo=FALSE, include=FALSE} -->
<!-- opts$min.weight <- 2 -->
<!-- metaccrna_filt = metaccrna[n>=opts$min.weight] -->
<!-- rm(metaccrna) -->
<!-- ``` -->

<!-- <!-- Rename annotations for plotting --> -->
<!-- ```{r preproc_data, echo=FALSE, include=FALSE, warning=FALSE} -->

<!-- anno_names <- c( -->
<!--   "prom_500_500_cgi"="CGI promoters", -->
<!--   "prom_500_500_noncgi"="non-CGI promoters", -->
<!--   "prom_500_500_active"="Active promoters", -->
<!--   "prom_500_500_inactive"="Inactive promoters", -->
<!--   "prom_500_500_poised"="Poised promoters", -->
<!--   "prom_500_500$"="Promoters", -->
<!--   "active_enhancers"="Active enhancers", -->
<!--   "super_enhancers"="Super enhancers", -->
<!--   "primed_enhancers"="Primed enhancers" -->
<!--   ) -->

<!-- to.plot$anno <- stringr::str_replace_all(to.plot$anno,anno_names) -->
<!-- ``` -->

```{r}
to.plot <- metacc %>% merge(sample_metadata[,c("sample","lineage")], by="sample")
```


<!-- Generate profiles stratified by lineage -->
```{r plot, echo=FALSE, include=TRUE}

f <- function(x) { return(data.frame(y=mean(x), ymin=mean(x)-sd(x), ymax=mean(x)+sd(x))) }

# to.plot[,tmp:=paste(context,expr_group,sep="_")] %>% .[,tmp:=factor(tmp, levels=c("CG_Low","CG_Medium","CG_High","GC_Low","GC_Medium","GC_High"))]

p_list <- list()
for (ann in unique(to.plot$anno)) {
  p <- ggplot(to.plot[anno==ann,], aes(x=dist, y=rate)) + 
    ggtitle(ann) +
    stat_summary(aes(group=interaction(lineage,context), colour=context, fill=context, linetype=lineage), fun.data=mean_se, geom="smooth", alpha=0.2, size=1.3) +
    xlab("Genomic distance from TSS") +
    ylab("Methylation/Accessibility rate") +
    # scale_linetype_manual(labels=c("2i","Serum"), values=c("dotted","solid")) +
    scale_color_manual(labels=c("CG methylation","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    scale_fill_manual(labels=c("CG methylation","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    # scale_color_manual(labels=c(" expression (CG)","GC accessibility"), values=c("#F8766D","#00BFC4")) +
    guides(
      colour=guide_legend(title="Context" ),
      linetype=guide_legend(override.aes=list(color="black"), title="Expression group"),
      fill=FALSE
      ) +
    theme(
      # plot.margin = unit(c(t=1,r=1,b=1,l=1), "cm"),
      plot.title = element_text(size=25,hjust=0.5),
      axis.text=element_text(size=rel(1.5), colour="black"),
      axis.title.x=element_text(size=20, margin=margin(10,0,0,0)),
      axis.title.y=element_text(size=20, margin=margin(0,10,0,0)),
      axis.line = element_line(size=rel(1.0)),
      axis.ticks = element_line(size=rel(1.2)),
      legend.key = element_blank(),
      legend.position = "right",
      legend.direction = "vertical",
      # legend.key.size= unit(0.5, "cm"),
      legend.key.width=unit(1.5,"line"),
      legend.key.height=unit(1.5,"line"),
      # legend.margin = margin(t=10, r=0, b=0, l=0, unit="pt"),
      legend.title = element_text(size=16, face="bold"),
      legend.text = element_text(size=18),
      panel.border=element_blank(),
      panel.grid.major=element_blank(),
      panel.grid.minor=element_blank(),
      panel.background=element_blank()
    )
  print(p)
  
  p_list[[ann]] <- p
  
  pdf(sprintf("%s/%s.pdf",io$outdir,ann), width=10, height=7, useDingbats=F)
  print(p)
  dev.off()
}
```


<!-- ```{r, echo=FALSE, include=FALSE} -->
<!-- pp <- cowplot::plot_grid(p_list[[2]],p_list[[3]],p_list[[4]],p_list[[5]], ncol=2, nrow=2, scale=0.97) -->
<!-- pdf(file=paste0(io$outdir,"/suppl.pdf"), width=18, height=12, useDingbats=F) -->
<!-- print(pp) -->
<!-- dev.off() -->
<!-- ``` -->

