---
title: "NMT-Seq ESC: Correlation between expression and methylation"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(data.table))
source("/Users/ricard/NMT-seq/correlations/utils.R")
```

```{r echo=FALSE, include=FALSE}
```

<!-- Define I/O and options -->
```{r echo=TRUE, include=FALSE}
# Data
io   <- list()
# io$base_dir   <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data"
# io$base_dir   <- "/Users/kapouranis/Documents/Andreas/Code/datasets/nmt-seq-data"
io$base_dir   <- "/Users/ricard/data/NMT-seq"
io$in.sample_metadata <- paste0(io$base_dir,"/sample_info.txt")
io$annos_dir  <- paste0(io$base_dir, "/features/filt")
io$rna_file   <- paste0(io$base_dir, "/rna/parsed/sceset.rds")
io$met_file   <- paste0(io$base_dir, "/met/parsed/met_data.tsv.gz")
io$outdir <- "/Users/ricard/NMT-seq/correlations/metrna/out"

# Filtering parameters
opts <- list()
opts$min.weight     <- 3      # Minimum weight per feature
opts$min.coverage   <- 0.4    # minimum coverage per feature across samples (met)
opts$min.cdr        <- 0.5    # minimum cellular detection rate per gene (expr)
opts$fraction.genes <- 0.5    # fraction of genes (expr) to keep based on variance
opts$fraction.sites <- 0.5    # fraction of sites (met) to keep based on variance
# opts$min.n          <- 100    # minimum number of features per annotation
opts$min.s          <- 20     # minimum number of samples to do the correlation
opts$gene_window    <- 8e3    # window length for the overlap between genes and features
opts$n_perms        <- 10     # Number of random permutations
opts$threshold_fdr  <- 0.10   # pvalue threshold for significance in FDR

opts$method <- "pearson"      # correlation type
opts$weight <- TRUE           # weighted correlation? 

opts$cor_samples <- F         # do correlation across cells?
opts$permutation <- F         # do permutation test?

opts$pseudobulk <- T          # Include pseudobulk estimates for the correlation across genes?

# Genomic contexts
opts$annos <- c("prom_2000_2000", "super_enhancers", "CTCF","genebody","p300","Oct4","DHS","Nanog","CGI","activeEnhancers_ENCODE")
# opts$annos <- c("activeEnhancers_ENCODE","activeEnhancers_Creyghton","prom_2000_2000","H3K27ac_Creyghton","H3K27ac_ENCODE", "H3K27me3_Creyghton", "H3K27me3_ENCODE", "H3K4me1_Creyghton", "H3K4me1_ENCODE", "H3K4me3_Creyghton", "H3K4me3_ENCODE")

# Define names for the genomic contexts
opts$anno_names <- c(
  "prom_2000_2000_cgi"="CGI promoters",
  "prom_2000_2000_noncgi"="non-CGI promoters",
  "prom_2000_2000$"="Promoters",
  "activeEnhancers_ENCODE"="Active enhancers",
  "super_enhancers"="Super enhancers",
  "primed_enhancers"="Primed enhancers",
  "exons"="Exons",
  "introns"="Introns",
  "genebody"="Gene body"
)

# Define which cells to use
opts$cells <- fread(io$in.sample_metadata, header=T) %>% .[pass_metQC==T & pass_rnaQC==T,sample]
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$in.sample_metadata, header=T) %>% .[sample%in%opts$cells]
```

<!-- Load RNA data -->
```{r load_rna, echo=FALSE, include=FALSE}
# Load expression as SCESet object
sceset <- readRDS(file = io$rna_file)

# Subset cells
sceset <- sceset[,opts$cells]
# sceset <- filter(sceset, culture=="Serum" & type=="NMT")

# Create data.table
rna <- exprs(sceset) %>% t %>% as.data.table(keep.rownames = "sample") %>% melt(id.vars = "sample", value.name = "expr", variable.name = "gene")

# Extract gene coordinates metadata to do the overlap
rna_metadata <- fData(sceset) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% .[,c("chr", "start", "end", "gene", "ens_id")]  %>% 
  .[, chr := as.factor(sub("chr", "", chr))] %>% setnames("ens_id", "id")
```

<!-- Load methylation data -->
```{r load_met, echo=FALSE, include=FALSE}
metrna_dt <- fread(sprintf("zcat < %s", io$met_file)) %>% as.data.table %>% .[anno%in%opts$annos & sample%in%opts$cells] %>% droplevels
```

<!-- Load genomic contexts -->
```{r load_genomiccontexts}
anno_dt <- lapply(opts$annos, function(anno) fread(sprintf("%s/%s.bed", io$annos_dir, anno), 
  colClasses = c("character", "integer", "integer", "character", "character", "character"))[, c(1, 2, 3, 4, 5, 6)]) %>%  
  rbindlist %>% setnames(c("chr", "start", "end", "strand", "id", "anno")) %>% .[,c("anno", "chr") := list(as.factor(anno), as.factor(chr))]
```

<!-- Associate the non-genic contexts with overlapping genes -->
```{r associate, echo=FALSE}
metrna_list <- list()
# Iterate over each feature
for (ann in unique(metrna_dt$anno)){
  # Subset corresponding anno
  met_tmp <- metrna_dt[anno == ann, ]
  
  # Non gene-associated feature
  if (all(grepl("ENSMUSG", unique(met_tmp$id)) == FALSE)) {
    # Extract coordiantes for methylation sites and for genes
    feature_metadata_tmp <- anno_dt[anno == ann, c("chr", "start", "end", "id")] %>% setkey(chr, start, end)
    rna_metadata_tmp <- rna_metadata[, c("chr", "start", "end", "gene")] %>%
      .[,c("start", "end") := list(start - opts$gene_window, end + opts$gene_window)] %>% setkey(chr, start, end)
    # Do the overlap
    ov <- foverlaps(feature_metadata_tmp, rna_metadata_tmp, nomatch = 0) %>% .[,c("gene", "id")]
    # Add expression levels
    ov_rna <- merge(rna, ov, by = "gene", allow.cartesian = TRUE)
    # Merge with methylation data
    metrna_list[[ann]] <- merge(ov_rna, met_tmp, by = c("sample", "id")) %>% 
      .[,c("sample", "id", "gene", "anno", "expr", "rate", "weight")]
  }
  # Gene-associated feature
  else if (all(grepl("ENSMUSG", unique(met_tmp$id)) == TRUE)) {
    met_tmp <- merge(met_tmp, rna_metadata[, c("id", "gene")], by = "id")
    metrna_list[[ann]] <- merge(rna, met_tmp, by = c("sample", "gene")) %>%
      .[,c("sample", "id", "gene", "anno", "expr", "rate", "weight")]
  }
}
metrna_dt <- rbindlist(metrna_list)
rm(metrna_list, met_tmp, feature_metadata_tmp, rna_metadata_tmp, ov, ov_rna)
```

<!-- Filter data to correlate across loci -->
```{r filter_all, echo=FALSE}
# Remove features with low weight (all types of correlation)
metrna_dt <- metrna_dt[weight >= opts$min.weight]
```

<!-- Filter data to correlate across cells -->
```{r filter_across_samples, echo=FALSE}

if (opts$cor_samples) {
  
  # To correlate across samples
  metrna_filt <- copy(metrna_dt)
  
  ## Remove genes/features with constant expression across samples
  keep_var_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[, .(var = var(expr)), by = "gene"] %>% .[var > 2, gene])
  keep_var_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[, .(var = wtd.var(rate, weight)), by = c("id")] %>% .[var > 2, id])
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(., names(.), function(x,y) x[gene %in% keep_var_genes[[y]] & id %in% keep_var_sites[[y]]]) %>% rbindlist
  
  ## Filter genes with low cellular detection rate and sites with low coverage across samples
  nsamples <- length(unique(metrna_filt$sample))
  keep_cov_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(cdr = sum(expr > 0) / .N), by = "gene"] %>% .[cdr >= opts$min.cdr, gene])
  keep_cov_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(cov = .N / nsamples), by = c("id")] %>% .[cov >= opts$min.coverage, id])
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(., names(.), function(x,y) x[gene %in% keep_cov_genes[[y]] & id %in% keep_cov_sites[[y]]]) %>% rbindlist
  
  ## Filter id-gene pairs with small number of samples (only required for the correlation across samples)
  keep_scov_sites <- metrna_filt[,.(n = .N), by = c("id", "anno")] %>% .[n >= opts$min.s] %>% .$id
  metrna_filt <- metrna_filt[id %in% keep_scov_sites]
  
  ## Extract highly variable genes/sites across samples
  keep_hv_genes <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(var = var(expr)), by = "gene"] %>% setorder(-var)  %>% head(n = nrow(.) * opts$fraction.genes) %>% .$gene)
  keep_hv_sites <- metrna_filt %>% split(.$anno) %>% map(~ .[,.(var = wtd.var(rate,weight)), by = c("id")] %>% setorder(-var)  %>% head(n = nrow(.) * opts$fraction.sites) %>% .$id)
  metrna_filt <- metrna_filt %>% split(.$anno) %>% map2(.,names(.), function(x,y) x[gene %in% keep_hv_genes[[y]] & id %in% keep_hv_sites[[y]]]) %>% rbindlist
  
  ## Filter annotations with few sites
  # keep_anno <- metrna_filt[, .(N = length(unique(id))), by = "anno"] %>% .[N > opts$min.n, anno]
  # metrna_filt <- metrna_filt[anno %in% keep_anno] %>% .[, anno := drop.levels(anno)]
  # keep_anno <- metrna_dt[, .(N = length(unique(id))), by = "anno"] %>% .[N > opts$min.n, anno]
  # metrna_dt <- metrna_dt[anno %in% keep_anno] %>% .[, anno := drop.levels(anno)]
  
}

```


<!-- Compute correlations  -->
```{r cor_rate, echo=FALSE, include=FALSE}
# Weighted correlation
if (opts$weight == TRUE){
  if (opts$method != "pearson") { print("Weighted correlation only supported for pearson"); stop() }
  # Correlate rate across samples
  if (opts$cor_samples) {
    cor_samples <- metrna_filt[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("id", "gene", "anno")]
  }
  # Correlate rate across genes
  cor_genes <- metrna_dt[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("sample", "anno")]
}else{
  if (opts$cor_samples) {
    # Correlate rate across samples
    cor_samples <- metrna_filt[, .(V1 = unlist(cor.test(rate, expr, alternative = "two.sided", method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("id", "gene", "anno")]
  }
  # Correlate rate across genes
  cor_genes <- metrna_dt[, .(V1 = unlist(cor.test(rate, expr, alternative = "two.sided", method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("sample", "anno")]
}

# Compute adjusted p-values across samples (both FDR and Bonferroni)
if (opts$cor_samples) {
  cor_samples <- cor_samples %>% .[, para := c("r", "t", "p")] %>% dcast(id + gene + anno ~ para, value.var = "V1") %>%
        .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
        .[, c("log_padj_fdr","log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
        .[, sig := padj_fdr <= opts$threshold_fdr] %>%  setorder(padj_fdr)
}

# Compute adjusted p-values across genes (both FDR and Bonferroni)
cor_genes <- cor_genes %>% .[,para := c("r", "t", "p")] %>% dcast(sample + anno ~ para, value.var = "V1") %>% 
      .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
      .[, c("log_padj_fdr", "log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
      .[, sig := padj_fdr <= opts$threshold_fdr] %>% setorder(padj_fdr)
```

<!-- Run permutation test for the correlation across samples -->
```{r perm_cor_rate, echo=FALSE, include=FALSE}

if (opts$cor_samples & opts$permutation) {
  pp_vals <- vector(mode = "numeric", length = length(cor_samples$p))
  # For each iteration...
  for (k in 1:opts$n_perms){
    print(k)
    # Copy original methylation data
    metrna_filt_perm <- copy(metrna_filt)
    # Permute gene expression levels
    metrna_filt_perm <- metrna_filt_perm[, expr := sample(expr), by = c("gene", "id", "anno")]
    # Compute correlation across samples
    # cor_samples_perm <- metrna_filt_perm[, .(p = wtd.cor(expr, rate, weight)[, c("p.value")]), by = c("id", "gene", "anno")]
    cor_samples_perm <- metrna_filt_perm[, .(p = cor.test(rate, expr, method=opts$method)[["p.value"]]), by = c("id","gene","anno")]
    # For each annotation sort by p-value
    cor_samples_perm <- cor_samples_perm %>% split(.$anno) %>% map(~ .[,.(anno = anno, p = sort(p))]) %>% rbindlist
    # Add the sorted p-values
    pp_vals <- pp_vals + cor_samples_perm %>% .[, p]
  }
  # Compute the average p-values
  pp_vals <- pp_vals / opts$n_perms
  # Store them in a data.table for later analysis
  cor_samples_perm <- cor_samples_perm[, p := pp_vals]
}

```

<!-- Rename annotations -->
```{r rename, echo=FALSE, include=FALSE}
cor_genes[,anno:=stringr::str_replace_all(cor_genes$anno, opts$anno_names)] %>% .[,anno:=as.factor(anno)]
if (opts$cor_samples) {
  cor_samples[,anno:=stringr::str_replace_all(cor_samples$anno, opts$anno_names)] %>% .[,anno:=as.factor(anno)]
  if (opts$permutation)
    cor_samples_perm[,anno:=stringr::str_replace_all(cor_samples_perm$anno, opts$anno_names)] %>% .[,anno:=as.factor(anno)]
}
```

<!-- Boxplot of correlations between methylation and expression -->
```{r boxplot_expr, echo=FALSE, include=TRUE, warning=FALSE, message=FALSE}

# Boxplot of correlations across samples
if (opts$cor_samples) {
  p <- ggplot(cor_samples, aes(x=anno, y=r)) +
    geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#F8766D") + geom_hline(yintercept=0, colour="black", linetype="dashed") +
    labs(title="Correlation across cells", x="", y="Weighted Pearson correlation coefficient") +
    scale_y_continuous(limits=c(-0.7,0.7)) + coord_flip() + boxplot_theme()
  print(p)
  # pdf(file=paste0(io$outdir, "/metrna_boxplot_acrosssamples.pdf"), width=8, height=6)
  # print(p)
  # dev.off()
}

# Boxplot of correlations across genes
p <- ggplot(cor_genes, aes(x=anno, y=r)) +
  geom_boxplot(alpha=0.8, outlier.shape=NA, fill="#F8766D") +
  geom_hline(yintercept=0, colour="black", linetype="dashed") +
  labs(title="", x="", y="Correlation coefficient") +
  # coord_flip(ylim=c(-0.9,0.9)) +
  coord_flip(ylim=c(-0.9,0.05)) +  
  boxplot_theme()
print(p)

# pdf(file=paste0(io$outdir, "/metrna_boxplot_acrossloci.pdf"), width=6, height=4)
# print(p)
# dev.off()

```

<!-- Pseudobulk correlation -->
```{r}

if (opts$pseudobulk) {
  
  metrna_pseudobulk_dt <- metrna_dt[,.(rate=round(mean(rate),2), expr=round(mean(expr),2), weight=.N),by=c("id","anno","gene")] %>% .[,sample:="pseudobulk"]
  
  cor_genes_pseudobulk <- metrna_pseudobulk_dt[, wtd.cor(rate, expr, weight)[, c("correlation", "t.value", "p.value")], by = c("sample", "anno")] %>%
    .[,para := c("r", "t", "p")] %>% dcast(sample + anno ~ para, value.var = "V1") %>% 
    .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
      .[, c("log_padj_fdr", "log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
      .[, sig := padj_fdr <= opts$threshold_fdr] %>% setorder(padj_fdr)
  
  cor_genes_pseudobulk[,anno:=stringr::str_replace_all(cor_genes_pseudobulk$anno, opts$anno_names)]
  
  p <- p + geom_point(aes(x=anno, y=r), color="black", fill="#F8766D", shape=21, size=3.0, data=cor_genes_pseudobulk)
  print(p)
}

```

<!-- Include bulk correlations estimates  -->
```{r}

opts$bulk <- T

if (opts$bulk) {
  cor_genes_bulk <- fread("/Users/ricard/NMT-seq/bulk_comparison/metrna/out/cor_metrna.txt") %>%
    .[anno%in%opts$annos] %>% 
    .[,anno:=stringr::str_replace_all(anno, opts$anno_names)] %>%
    .[,anno:=factor(anno,levels=levels(cor_genes$anno))]
    
  p <- p + geom_point(aes(x=anno, y=r), color="black", fill="#F8766D", shape=8, size=3.0, data=cor_genes_bulk)
  print(p)
  
  pdf(file=paste0(io$outdir, "/metrna_boxplot_acrossloci_witbulk.pdf"), width=6, height=5, useDingbats = F)
  print(p)
  dev.off()
}

```


<!-- Volcano plot of p-values against Pearson's r and QQ-plot of p-values -->
```{r, echo=FALSE}
if (opts$cor_samples & opts$permutation) {
  pp <- qq <- list()
  for (n in unique(cor_samples$anno)) {
    
    # Compute positive and negative hits in terms of correlation
    negative_hits <- cor_samples[anno==n & sig==TRUE & r<0,id]
    positive_hits <- cor_samples[anno==n & sig==TRUE & r>0,id]
    all <- nrow(cor_samples[anno == n,])
    
    # Generate volcano plot  
    pp[[n]] <- gg_volcano_plot(cor_samples[anno == n,], title = "")
    
    # Generate permutation plot
    qq[[n]] <- gg_qqplot(cor_samples[anno == n, ], cor_samples_perm[anno == n]$p, title = "")
    
    # Combine and save
    print(plot_grid(pp[[n]], qq[[n]], labels = c("", ""), label_size = 20, ncol = 2, nrow = 1))
    pdf(file=paste0(io$outdir,"/volcano_", n, ".pdf"), width = 14, height = 8, useDingbats = FALSE)
    grid.arrange(pp[[n]], qq[[n]], ncol=2, top = textGrob(n, gp=gpar(fontsize=29, fontface = "bold")), newpage = TRUE)
    dev.off()
    
  }
}
```



<!-- <!-- Scatterplot of correlations across cells --> -->
<!-- ```{r scatterplots, echo=FALSE, include=FALSE} -->

<!-- # anno_foo="prom_2000_2000" -->
<!-- # sample_foo="E01" -->
<!-- anno_foo="super_enhancers" -->
<!-- sample_foo="B12" -->
<!-- p <- ggplot(metrna_dt[anno==anno_foo & sample==sample_foo], aes(x=rate,y=expr)) + -->
<!--   geom_point(alpha=0.5) + -->
<!--   stat_smooth(method="lm", color="black", alpha=0.5) + -->
<!--   # facet_wrap(~sample) + -->
<!--   xlab("Methylation rate") + ylab("Expression") + -->
<!--   theme( -->
<!--   plot.title = element_text(size=30, hjust=0.5, margin=margin(0,0,20,0)), -->
<!--   axis.title.y = element_text(colour="black", size=28, vjust=1.5), -->
<!--   axis.title.x = element_text(colour="black", size=28, vjust=1.5, margin=margin(15,0,0,0)), -->
<!--   axis.text.x = element_text(colour="black",size=rel(1.8)), -->
<!--   axis.text.y = element_text(colour="black",size=rel(1.8)), -->
<!--   # axis.line = element_line(colour="black", size=rel(0.7)), -->
<!--   # axis.ticks.x = element_line(colour="black", size=rel(0.8)), -->
<!--   # axis.ticks.y = element_blank(), -->
<!--   legend.position="none", -->
<!--   panel.background = element_blank(), -->
<!--   panel.border = element_blank(), -->
<!--   panel.grid = element_blank() -->
<!--   ) -->

<!-- # pdf(file=paste0("out/scatter_", features[i], ".pdf"), width = 8, height = 8) -->
<!-- # print(p) -->
<!-- # dev.off() -->

<!-- ``` -->


<!-- Save results -->
```{r}
fwrite(round_df(cor_genes,5), paste0(io$outdir,"/metrna_cor_loci.tsv"), sep="\t")
if (opts$cor_samples) {
  fwrite(round_df(cor_samples,5), paste0(io$outdir,"/metrna_cor_samples.tsv"), sep="\t")
}
```
