---
title: "NMT-seq serum cells: general statistics of genomic features"
output: 
  BiocStyle::html_document: 
    toc: false
    fig_width: 7
    fig_height: 7
---

```{r load_modules, echo=FALSE, include=FALSE}
library(weights)
library(data.table)
library(purrr)
library(ggplot2)
```

```{r, echo=FALSE}

# sample_colors <- function(n) {
#   hues = seq(15, 375, length = n + 1)
#   hcl(h = hues, l = 65, c = 100)[1:n]
# }

boxplot_theme <- function() {
  p <- theme(
    plot.title = element_blank(),
    axis.title.y = element_blank(),
    axis.title.x = element_text(colour="black", size=20, vjust=1.5, margin=margin(10,0,0,0)),
    axis.text.x = element_text(colour="black",size=rel(1.7)),
    axis.text.y = element_text(colour="black",size=rel(1.7)),
    axis.line = element_line(colour="black", size=rel(0.7)),
    axis.ticks.x = element_line(colour="black", size=rel(1.0)),
    axis.ticks.y = element_line(colour="black", size=rel(1.0)),
    # axis.ticks.y = element_blank(),
    panel.background = element_blank(),
    panel.grid = element_blank(),
    legend.position="top",
    legend.text=element_text(size=15),
    legend.title=element_blank(),
    legend.background=element_blank(),
    panel.border = element_blank()
  )
}
```

```{r define_options, echo=FALSE}

## I/O ##
io <- list()
io$met.infile <- "/Users/ricard/data/NMT-seq/met/parsed/met_data.tsv.gz"
io$acc.infile <- "/Users/ricard/data/NMT-seq/acc/parsed/acc_data.tsv.gz"
io$metadata <- "/Users/ricard/data/NMT-seq/sample_info.txt"
io$outdir <- "/Users/ricard/NMT-seq/stats/features/out"

## Options ##
opts <- list()

# Define which cells to use
opts$acc_cells <- fread(io$metadata) %>% .[pass_accQC==T,sample]
opts$met_cells <- fread(io$metadata) %>% .[pass_metQC==T,sample]
    
    
# Define which annotations to use and the order for plotting
opts$acc_annos <- c("prom_2000_2000_cgi","prom_2000_2000_noncgi","exons","introns","active_enhancers","super_enhancers","primed_enhancers","CTCF","Nanog","Oct4","IAP","DHS","CGI","p300")
opts$met_annos <- c("prom_2000_2000_cgi","prom_2000_2000_noncgi","exons","introns","active_enhancers","super_enhancers","primed_enhancers","CTCF","Nanog","Oct4","IAP","DHS","CGI","p300")

# Filter features with low coverage
# opts$min_weight <- 1
# opts$min_cov <- 0.5

opts$anno_names <- c(
  "prom_2000_2000_cgi"="CGI promoters",
  "prom_2000_2000_noncgi"="non-CGI promoters",
  "genebody"="Gene body",
  "exons"="Exons",
  "introns"="Introns",
  "active_enhancers"="Active enhancers",
  "super_enhancers"="Super enhancers",
  "primed_enhancers"="Primed enhancers",
  "prom_2000_2000_cgi"="CGI promoters",
  "prom_2000_2000_noncgi"="non-CGI promoters"
)

opts$anno_order <- c("CGI promoters","non-CGI promoters", "Exons", "Introns", "Active enhancers", "Super enhancers", "Primed enhancers", "CTCF", "DHS", "p300", "Nanog", "Oct4","IAP","CGI")

```


<!-- Load sample metadata -->
```{r load_metadata}
metadata <- fread(io$metadata) %>% .[sample %in% union(opts$met_cells,opts$acc_cells)]
```

<!-- Load methylation and accessibility data -->
```{r load_metacc, echo=FALSE, include=FALSE}
met <- fread(sprintf("zcat < %s",io$met.infile), showProgress=F) %>% .[anno%in%opts$met_annos & sample%in%opts$met_cells] %>% .[,anno:=factor(anno, levels=opts$met_annos)] %>% .[,context:="CG"]
acc <- fread(sprintf("zcat < %s",io$acc.infile), showProgress=F) %>% .[anno%in%opts$acc_annos & sample%in%opts$acc_cells] %>% .[,anno:=factor(anno, levels=opts$acc_annos)] %>% .[,context:="GC"]
metacc <- rbind(met,acc)

# Merge with metadata
metacc <- merge(metadata[,c("sample","culture")], metacc, by="sample")
```

```{r calculate_stats, echo=FALSE, include=FALSE}
feature_stats <- metacc[,.(mean=mean(rate, na.rm=T), var=var(rate, na.rm=T)), by=c("anno","id","context")] %>%
  .[,anno:=stringr::str_replace_all(anno, opts$anno_names)] %>%
  .[,anno:=factor(anno,levels = opts$anno_order)] %>% droplevels

```

<!-- Mean methylation/accessibility rate per genomic context -->
```{r mean, echo=FALSE, include=TRUE}
p1 <- ggplot(feature_stats[context=="CG",c("anno","mean")], aes(x=anno, y=mean)) +
  geom_boxplot(fill="#F8766D", alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Mean methylation rate") +
  coord_flip() +
  # scale_y_continuous(limits=c(0,100)) +
  boxplot_theme()
p1

p2 <- ggplot(feature_stats[context=="GC",c("anno","mean")], aes(x=anno, y=mean)) +
  geom_boxplot(fill="#00BFC4", alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Mean accessibility rate") +
  # scale_y_continuous(limits=c(0,100)) +
  coord_flip() +
  boxplot_theme()
p2

# pdf("/Users/ricard/NMT-seq/stats/features/out/boxplot_met_mean.pdf", height=5, width=5)
# print(p1)
# dev.off()
# 
# pdf("/Users/ricard/NMT-seq/stats/features/out/boxplot_acc_mean.pdf", height=5, width=5)
# print(p2)
# dev.off()

```

<!-- Variance across cells on methylation/accessibility rate per genomic context -->
```{r met_var, echo=FALSE, include=TRUE, warning=FALSE}

p3 <- ggplot(feature_stats[context=="CG",c("anno","var")], aes(x=anno, y=var)) +
  geom_boxplot(fill="#F8766D", alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Cell-to-cell variance on the methylation rate") +
  coord_flip(ylim=c(0,2500)) +
  boxplot_theme()
p3

p4 <- ggplot(feature_stats[context=="GC",c("anno","var")], aes(x=anno, y=var)) +
  geom_boxplot(fill="#00BFC4", alpha=0.8, coef=0, outlier.shape=NA) +
  ggtitle("") + xlab("") + ylab("Cell-to-cell variance on the accessibility rate") +
  coord_flip(ylim=c(0,2000)) +
  # scale_y_continuous(limits=c(0,1400)) +
  boxplot_theme()
p4


# pdf("/Users/ricard/NMT-seq/stats/features/out/boxplot_met_var.pdf", height=10, width=10)
# print(p1)
# dev.off()
# 
# pdf("/Users/ricard/NMT-seq/stats/features/out/boxplot_acc_var.pdf", height=10, width=10)
# print(p2)
# dev.off()
```

```{r}
pdf("/Users/ricard/NMT-seq/stats/features/out/boxplots.pdf", height=12, width=17)
print(cowplot::plot_grid(p1,p3,p2,p4, labels = c("a","b","c","d"), label_size=20, ncol=2, nrow=2))
dev.off()
```


<!-- Scatterplots of CG mean/var vs GC mean/var -->
```{r scatterplot, echo=FALSE}

# opts$anno_colors <- sample_colors(n=length(unique(opts$anno_names)))
# names(opts$anno_colors) <- c("CGI promoters","non-CGI promoters","Exons","Introns","Active enhancers","Super enhancers","Primed enhancers","Nanog","Oct4","p300","IAP","DHS", "CTCF")

tmp <- dcast(feature_stats, anno+id~context, value.var=c("mean","var")) %>% 
  .[,.(mean_CG=mean(mean_CG,na.rm=T), mean_GC=mean(mean_GC,na.rm=T), var_CG=mean(var_CG,na.rm=T), var_GC=mean(var_GC,na.rm=T)), by=c("anno")]


p1 <- ggplot(tmp, aes(x=mean_CG, y=mean_GC, color=anno)) + 
  geom_point(size=2.5) +
  # geom_text(aes(label=anno), hjust=0, vjust=0) +
  stat_smooth(method="lm", color="black", linetype="dashed", size=0.5, alpha=0.1) +
  ggrepel::geom_text_repel(aes(label=anno), size=8, force=10) +
  # scale_x_continuous(limits=c(0,100)) +
  # scale_y_continuous(limits=c(0,100)) +
  xlab("Mean methylation rate") + ylab("Mean accessibility rate") +
  theme(
    plot.margin=margin(10,10,10,10),
    axis.text.x = element_text(colour="black",size=rel(1.8)),
    axis.text.y = element_text(colour="black",size=rel(1.8)),
    axis.title.x = element_text(colour="black",size=20, margin=margin(10,0,0,0)),
    axis.title.y = element_text(colour="black",size=20, margin=margin(0,10,0,0)),
    axis.ticks.y = element_line(colour="black", size=rel(1.3)),
    axis.line = element_line(colour="black"),
    legend.position="none",
    panel.background = element_blank()
  )
p1

p2 <- ggplot(tmp, aes(x=mean_CG, y=var_GC, color=anno)) +
  geom_point(size=2.5) +
  # geom_text(aes(label=anno), hjust=0, vjust=0) +
  ggrepel::geom_text_repel(aes(label=anno), size=8, force=10) +
  stat_smooth(method="lm", color="black", linetype="dashed", size=0.5, alpha=0.1) +
  xlab("Variance on methylation rate") + ylab("Variance on accessibility rate") +
  theme(
    plot.margin=margin(10,10,10,10),
    axis.text.x = element_text(colour="black",size=rel(1.8)),
    axis.text.y = element_text(colour="black",size=rel(1.8)),
    axis.title.x = element_text(colour="black",size=20, margin=margin(10,0,0,0)),
    axis.title.y = element_text(colour="black",size=20, margin=margin(0,10,0,0)),
    axis.ticks.y = element_line(colour="black", size=rel(1.3)),
    axis.line = element_line(colour="black"),
    legend.position="none",
    panel.background = element_blank()
  )
p2

pdf("/Users/ricard/NMT-seq/stats/features/out/scatterplot_accmet_mean.pdf", height=6, width=6, useDingbats=FALSE)
print(p1)
dev.off()

pdf("/Users/ricard/NMT-seq/stats/features/out/scatterplot_accmet_var.pdf", height=6, width=6, useDingbats=FALSE)
print(p2)
dev.off()
```


<!-- ## Total number of features per genomic context -->
<!-- ```{r met_total_features, echo=FALSE, include=TRUE} -->
<!-- tmp <- feature_stats[,.N,by=c("anno","context")] %>% .[,fill:=anno] -->
<!-- p <- ggplot(tmp, aes(x=anno,y=N,fill=context)) + -->
<!--   geom_bar(stat='identity', position="dodge") + -->
<!--   xlab("") + ylab("Number of features") + -->
<!--   coord_flip() + -->
<!--   scale_y_continuous(expand=c(0,0)) + -->
<!--   scale_fill_discrete(labels = c("CG methylation","GC accessibility")) + -->
<!--   theme( -->
<!--     axis.text.x = element_text(colour="black",size=rel(1.5)), -->
<!--     axis.text.y = element_text(colour="black",size=rel(1.5), vjust=0.5), -->
<!--     axis.title.x = element_text(colour="black",size=15), -->
<!--     axis.ticks.y = element_line(colour="black", size=rel(1.2)), -->
<!--     legend.position="top", -->
<!--     legend.title = element_blank(), -->
<!--     legend.direction = "horizontal", -->
<!--     legend.key.width=unit(1.2,"line"), -->
<!--     legend.key.height=unit(1.0,"line"), -->
<!--     legend.text = element_text(size=15), -->
<!--     panel.background = element_blank(), -->
<!--     panel.grid = element_blank(), -->
<!--     panel.border = element_blank() -->
<!--   ) -->
<!-- p -->
<!-- knitr::kable( tmp %>% dcast(anno~context,value.var="N") ) -->
<!-- ``` -->





